
### Note: folder spades_assembly_all_illumina is in the lab0809 directory with the complete structure of lab08 output, but is not uploaded onto github due to the file size limitation

## Part 1: Generate alignment with SPAdes
    
`spades.py -1 /data/lab8/illumina_reads_R1.fastq.gz -2 /data/lab8/illumina_reads_R2.fastq --pacbio /data/lab8/pacbio_reads.fastq --careful --cov-cutoff auto -o spades_assembly_all_illumina -t 1`
    
left on screen starting 10/11/18 5:10pm PST

output in `cd ~/fa18-BioE131/lab08/spades_assembly_all_illumina`

Question:
- Why do we expect short reads to produce a more fragmented assembly than long reads? 
> In general, more short reads are needed to produce the same coverage as a long reads data. The contigs are shorter and there're many contigs, also there may be regions that are not being covered in the reads due to the short read length. For example, the repetitive regions may be considered as small fragments because the read length is unable to detect the larger patterns. All of the issues above lead to a more fragmented they long reads. Also, if the sequence is de novo and so a reference doesn't exist, repeated areas can cause a lot of difficulty in sequence assembly. Additional difficulties include base substitutions by inaccurate polymerases, chimeric sequences, and PCR-bias, all of which can contribute to generating an incorrect sequence.

- Why does a single-molecule sequencing like PacBio have a higher error rate than Illumina?  
> As PacBio requencing differs significantly from the Illumina sequencing in that there is no need for amplification during the library preparation, nor during the sequencing process, and it apply a movie capture of the polymerase activity, the signal to noise ratio is larger and so the error rate is also higher, as both mismatches and insertion proportion is larger than in a illunina sequencing.

## Assembly-stats
- Using the assembly-statsprogram, calculate statistics on both your contigs and scaffolds file
- Report the total length of all contigs (or scaffolds), the number of contigs (or scaffolds), and the N50 

Code Line: output a tabular analysis with headers
`assembly-stats -t ./spades_assembly_all_illumina/scaffolds.fasta ./spades_assembly_all_illumina/contigs.fasta > ./assembly_stats_report.txt`


```python
import pandas as pd
data = pd.read_csv('assembly_stats_report.csv',sep = "	")
data
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>filename</th>
      <th>total_length</th>
      <th>number</th>
      <th>mean_length</th>
      <th>longest</th>
      <th>shortest</th>
      <th>N_count</th>
      <th>Gaps</th>
      <th>N50</th>
      <th>N50n</th>
      <th>N70</th>
      <th>N70n</th>
      <th>N90</th>
      <th>N90n</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>./spades_assembly_all_illumina/scaffolds.fasta</td>
      <td>6658426</td>
      <td>143</td>
      <td>46562.42</td>
      <td>537743</td>
      <td>128</td>
      <td>144</td>
      <td>7</td>
      <td>332489</td>
      <td>8</td>
      <td>207869</td>
      <td>14</td>
      <td>92098</td>
      <td>24</td>
    </tr>
    <tr>
      <th>1</th>
      <td>./spades_assembly_all_illumina/contigs.fasta</td>
      <td>6654856</td>
      <td>149</td>
      <td>44663.46</td>
      <td>489856</td>
      <td>128</td>
      <td>0</td>
      <td>0</td>
      <td>230325</td>
      <td>10</td>
      <td>168584</td>
      <td>16</td>
      <td>81820</td>
      <td>28</td>
    </tr>
  </tbody>
</table>
</div>




```python
df1 = data[['filename','total_length',"number","N50"]]
df1
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>filename</th>
      <th>total_length</th>
      <th>number</th>
      <th>N50</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>./spades_assembly_all_illumina/scaffolds.fasta</td>
      <td>6658426</td>
      <td>143</td>
      <td>332489</td>
    </tr>
    <tr>
      <th>1</th>
      <td>./spades_assembly_all_illumina/contigs.fasta</td>
      <td>6654856</td>
      <td>149</td>
      <td>230325</td>
    </tr>
  </tbody>
</table>
</div>



- Why is N50 is useful statistic to calculate?  
> N50 is useful in that if evaluates the assembly quality in terms of contiguity, as it's defined as the senquence length of the shortest contig at 50% of the total genome length. It is useful to generate the distributive analysis and a half cut in terms of contig.
- Why not just list the mean or median contig length?   
> The mean and median contig length does not take into account of the contiguity and the total length of the read length being assemblied. In that way, the mean and median is not a even cut point in terms of the number of bases
- In your assembly, are scaffolds longer than contigs, or are scaffolds approximately equal to contigs?  
> Scaffolds are longer than contigs, around 50% longer N50 and much longer in terms of longest and mean segments

## Extra Credit:
- Extract the coverage from each FASTA header
code: apply grep and awk, use ">" to separate each fasta, "_" to separate names, and print the name, length and coverage field
`grep '^>'  contigs.fasta | awk -F _  'BEGIN {OFS="\t"} {print $0,$4,$6}'`
- plot a histogram of coverage for all contigs in your assembly


```python
import pandas as pd
data = pd.read_csv('contig_cov.csv',sep = "	", header = None,names = ["name", "length","Coverage"])
# A demonstration of the first five lines
print(data[1:5])
# extract the second column as the coverage
cov = data[["Coverage"]]
cov.hist(bins=149)

# removing the single extremely high anomaly (multiple G and multiple T)for a closer look
cov2 = data[0:147][["Coverage"]]
cov2.hist(bins=140)
```

                                     name  length  Coverage
    1  >NODE_2_length_468307_cov_8.954830  468307  8.954830
    2  >NODE_3_length_460222_cov_8.835066  460222  8.835066
    3  >NODE_4_length_432255_cov_8.768225  432255  8.768225
    4  >NODE_5_length_352739_cov_8.419214  352739  8.419214





    array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7ff1c502cb00>]],
          dtype=object)




![png](output_8_2.png)



![png](output_8_3.png)


- Is coverage uniformly distributed? Does it look Gaussian? Is it bimodal or trimodal? 
> The coverage is not uniformly distributed, and has a closest look of bimodal  
- What explains the presence of contigs with coverage an integer multiple of the mean coverage?
> mean coverage = total mapped reads time/total reference length, and so the contig coverage = K * total mapped reads time/total reference length. A possible explanation is that the contig length is a 1/k of the total length, where k is a positive integer larger than 1, also the read time will be a 1/k' of the total read times. If both two condition is satisfied, a contig converage could be a multiple of the mean coverage.


## Identify the taxonfrom which your genome originated
- “hypervariable regions”can be used to determine the taxonfrom which a 16S rRNA gene originated  
1) scan over the entire genome to locate copies of the 16S ribosomal RNA gene  
2) extract the 16S rRNA genes from your assembly   
3) compare these genes to a database of known 16S rRNA genes.

Terminal Code:
`rna_hmm3.py -i ./spades_assembly_all_illumina/contigs.fasta -o ./ran_hmm_out.txt`   
Inspect the output


```python
import pandas as pd
data = pd.read_csv('ran_hmm_out.txt',sep = "	")
data
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>##seq_name</th>
      <th>method</th>
      <th>feature</th>
      <th>start</th>
      <th>end</th>
      <th>evalue</th>
      <th>strand</th>
      <th>frame</th>
      <th>attribute</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>NODE_5_length_352739_cov_8.419214</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>65159</td>
      <td>68063</td>
      <td>0.000000e+00</td>
      <td>-</td>
      <td>NaN</td>
      <td>23S_rRNA</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NODE_21_length_103352_cov_10.011827</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>37580</td>
      <td>40484</td>
      <td>0.000000e+00</td>
      <td>+</td>
      <td>NaN</td>
      <td>23S_rRNA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NODE_1_length_489856_cov_8.332587</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>457</td>
      <td>3361</td>
      <td>0.000000e+00</td>
      <td>+</td>
      <td>NaN</td>
      <td>23S_rRNA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NODE_2_length_468307_cov_8.954830</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>161099</td>
      <td>164003</td>
      <td>0.000000e+00</td>
      <td>+</td>
      <td>NaN</td>
      <td>23S_rRNA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NODE_98_length_451_cov_1.006173</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>184</td>
      <td>303</td>
      <td>6.500000e-36</td>
      <td>+</td>
      <td>NaN</td>
      <td>23S_rRNA</td>
    </tr>
    <tr>
      <th>5</th>
      <td>NODE_98_length_451_cov_1.006173</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>147</td>
      <td>220</td>
      <td>1.400000e-09</td>
      <td>-</td>
      <td>NaN</td>
      <td>23S_rRNA</td>
    </tr>
    <tr>
      <th>6</th>
      <td>NODE_13_length_195508_cov_8.764542</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>49</td>
      <td>1592</td>
      <td>0.000000e+00</td>
      <td>-</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>7</th>
      <td>NODE_2_length_468307_cov_8.954830</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>159050</td>
      <td>160594</td>
      <td>0.000000e+00</td>
      <td>+</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>8</th>
      <td>NODE_5_length_352739_cov_8.419214</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>68568</td>
      <td>70112</td>
      <td>0.000000e+00</td>
      <td>-</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>9</th>
      <td>NODE_32_length_47877_cov_11.378513</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>46286</td>
      <td>47829</td>
      <td>0.000000e+00</td>
      <td>+</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>10</th>
      <td>NODE_21_length_103352_cov_10.011827</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>35531</td>
      <td>37075</td>
      <td>0.000000e+00</td>
      <td>+</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>11</th>
      <td>NODE_21_length_103352_cov_10.011827</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>40649</td>
      <td>40763</td>
      <td>2.900000e-21</td>
      <td>+</td>
      <td>NaN</td>
      <td>5S_rRNA</td>
    </tr>
    <tr>
      <th>12</th>
      <td>NODE_19_length_133461_cov_8.643403</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>133332</td>
      <td>133446</td>
      <td>3.800000e-21</td>
      <td>-</td>
      <td>NaN</td>
      <td>5S_rRNA</td>
    </tr>
    <tr>
      <th>13</th>
      <td>NODE_5_length_352739_cov_8.419214</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>64880</td>
      <td>64994</td>
      <td>1.000000e-20</td>
      <td>-</td>
      <td>NaN</td>
      <td>5S_rRNA</td>
    </tr>
    <tr>
      <th>14</th>
      <td>NODE_40_length_25337_cov_8.961166</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>2</td>
      <td>108</td>
      <td>1.600000e-20</td>
      <td>+</td>
      <td>NaN</td>
      <td>5S_rRNA</td>
    </tr>
    <tr>
      <th>15</th>
      <td>NODE_1_length_489856_cov_8.332587</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>3528</td>
      <td>3642</td>
      <td>1.500000e-20</td>
      <td>+</td>
      <td>NaN</td>
      <td>5S_rRNA</td>
    </tr>
    <tr>
      <th>16</th>
      <td>NODE_2_length_468307_cov_8.954830</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>164170</td>
      <td>164284</td>
      <td>1.400000e-20</td>
      <td>+</td>
      <td>NaN</td>
      <td>5S_rRNA</td>
    </tr>
  </tbody>
</table>
</div>



## delete all lines within other than those containing 16S_rRNA genes


```python
import pandas as pd
data = pd.read_csv('ran_hmm_out.GFF',sep = "	")
data
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>##seq_name</th>
      <th>method</th>
      <th>feature</th>
      <th>start</th>
      <th>end</th>
      <th>evalue</th>
      <th>strand</th>
      <th>frame</th>
      <th>attribute</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>NODE_13_length_195508_cov_8.764542</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>49</td>
      <td>1592</td>
      <td>0</td>
      <td>-</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NODE_2_length_468307_cov_8.954830</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>159050</td>
      <td>160594</td>
      <td>0</td>
      <td>+</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NODE_5_length_352739_cov_8.419214</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>68568</td>
      <td>70112</td>
      <td>0</td>
      <td>-</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NODE_32_length_47877_cov_11.378513</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>46286</td>
      <td>47829</td>
      <td>0</td>
      <td>+</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NODE_21_length_103352_cov_10.011827</td>
      <td>rna_hmm3</td>
      <td>rRNA</td>
      <td>35531</td>
      <td>37075</td>
      <td>0</td>
      <td>+</td>
      <td>NaN</td>
      <td>16S_rRNA</td>
    </tr>
  </tbody>
</table>
</div>



## use bedtools getfastato extract nucleic acid sequences of the 16S rRNA genes from your assembly
- specify the path to contigs.fasta with -fiand the path to the GFF file you obtained above, with -bed
- The output will be in FASTA format.
`bedtools getfasta -fi ./spades_assembly_all_illumina/contigs.fasta -bed ./ran_hmm_out.GFF -fo 16s_rRNA.fasta`
> The output is in fasta format named 16s_rRNA.fasta. For inspection:  
>NODE_13_length_195508_cov_8.764542:48-1592
CGATTAAGGAGGTGATCCAGCCGCAGGTTCCCCTACGGCTACCTTGTTACGACTTCACCCCAGTCATGAATCACACCGTGGTAACCGTCCTCCCGAAGGTTAGACTAGCTACTTCTGGTGCAACCCACTCCCATGGTGTGACGGGCGGTGTGTACAAGGCCCGGGAACGTATTCACCGCGACATTCTGATTCGCGATTACTAGCGATTCCGACTTCACGCAGTCGAGTTGCAGACTGCGATCCGGACTACGATCGGTTTTCTGGGATTAGCTCCACCTCGCGGCTTGGCAACCCTCTGTACCGACCATTGTAGCACGTGTGTAGCCCAGGCCGTAAGGGCCATGATGACTTGACGTCATCCCCACCTTCCTCCGGTTTGTCACCGGCAGTCTCCTTAGAGTGCCCACCATTACGTGCTGGTAACTAAGGACAAGGGTTGCGCTCGTTACGGGACTTAACCCAACATCTCACGACACGAGCTGACGACAGCCATGCAGCACCTGTCTCAATGTTCCCGAAGGCACCAATCCATCTCTGGAAAGTTCATTGGATGTCAAGGCCTGGTAAGGTTCTTCGCGTTGCTTCGAATTAAACCACATGCTCCACCGCTTGTGCGGGCCCCCGTCAATTCATTTGAGTTTTAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTAATGCGTTAGCTGCGCCACTAAGAGCTCAAGGCTCCCAACGGCTAGTTGACATCGTTTACGGCGTGGACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCACCTCAGTGTCAGTATCAGTCCAGGTGGTCGCCTTCGCCACTGGTGTTCCTTCCTATATCTACGCATTTCACCGCTACACAGGAAATTCCACCACCCTCTACCATACTCTAGCTCGTCAGTTTTGAATGCAGTTCCCAGGTTGAGCCCGGGGATTTCACATCCAACTTAACGAACCACCTACGCGCGCTTTACGCCCAGTAATTCCGATTAACGCTTGCACCCTCTGTATTACCGCGGCTGCTGGCACAGAGTTAGCCGGTGCTTATTCTGTCGGTAACGTCAAAACAATTACGTATTAGGTAACTGCCCTTCCTCCCAACTTAAAGTGCTTTACAATCCGAAGACCTTCTTCACACACGCGGCATGGCTGGATCAGGCTTTCGCCCATTGTCCAATATTCCCCACTGCTGCCTCCCGTAGGAGTCTGGACCGTGTCTCAGTTCCAGTGTGACTGATCATCCTCTCAGACCAGTTACGGATCGTCGCCTTGGTGAGCCATTACCTCACCAACTAGCTAATCCGACCTAGGCTCATCTGATAGCGCAAGGCCCGAAGGTCCCCTGCTTTCTCCCGTAGGACGTATGCGGTATTAGCGTCCGTTTCCGAGCGTTATCCCCCACTACCAGGCAGATTCCTAGGCATTACTCACCCGTCCGCCGCTCTCAAGAGGTGCAAGCACCTCTCTACCGCTCGACTTGCATGTGTTAGGCCTGCCGCCAGCGTTCAATCTGAGCCATGATCAAACTCTTCAGTTCAAA
>NODE_2_length_468307_cov_8.954830:159049-160594
TTTGAACTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAGAGAGGTGCTTGCACCTCTTGAGAGCGGCGGACGGGTGAGTAATGCCTAGGAATCTGCCTGGTAGTGGGGGATAACGCTCGGAAACGGACGCTAATACCGCATACGTCCTACGGGAGAAAGCAGGGGACCTTCGGGCCTTGCGCTATCAGATGAGCCTAGGTCGGATTAGCTAGTTGGTGAGGTAATGGCTCACCAAGGCGACGATCCGTAACTGGTCTGAGAGGATGATCAGTCACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGGACAATGGGCGAAAGCCTGATCCAGCCATGCCGCGTGTGTGAAGAAGGTCTTCGGATTGTAAAGCACTTTAAGTTGGGAGGAAGGGCAGTTACCTAATACGTAATTGTTTTGACGTTACCGACAGAATAAGCACCGGCTAACTCTGTGCCAGCAGCCGCGGTAATACAGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTCGTTAAGTTGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATTCAAAACTGACGAGCTAGAGTATGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAAATGCGTAGATATAGGAAGGAACACCAGTGGCGAAGGCGACCACCTGGACTGATACTGACACTGAGGTGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCAACTAGCCGTTGGGAGCCTTGAGCTCTTAGTGGCGCAGCTAACGCATTAAGTTGACCGCCTGGGGAGTACGGCCGCAAGGTTAAAACTCAAATGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGCCTTGACATCCAATGAACTTTCCAGAGATGGATTGGTGCCTTCGGGAACATTGAGACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGTAACGAGCGCAACCCTTGTCCTTAGTTACCAGCACGTAATGGTGGGCACTCTAAGGAGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAGTCATCATGGCCCTTACGGCCTGGGCTACACACGTGCTACAATGGTCGGTACAGAGGGTTGCCAAGCCGCGAGGTGGAGCTAATCCCAGAAAACCGATCGTAGTCCGGATCGCAGTCTGCAACTCGACTGCGTGAAGTCGGAATCGCTAGTAATCGCGAATCAGAATGTCGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCATGGGAGTGGGTTGCACCAGAAGTAGCTAGTCTAACCTTCGGGAGGACGGTTACCACGGTGTGATTCATGACTGGGGTGAAGTCGTAACAAGGTAGCCGTAGGGGAACCTGCGGCTGGATCACCTCCTTAATCGA
>NODE_5_length_352739_cov_8.419214:68567-70112
TCGATTAAGGAGGTGATCCAGCCGCAGGTTCCCCTACGGCTACCTTGTTACGACTTCACCCCAGTCATGAATCACACCGTGGTAACCGTCCTCCCGAAGGTTAGACTAGCTACTTCTGGTGCAACCCACTCCCATGGTGTGACGGGCGGTGTGTACAAGGCCCGGGAACGTATTCACCGCGACATTCTGATTCGCGATTACTAGCGATTCCGACTTCACGCAGTCGAGTTGCAGACTGCGATCCGGACTACGATCGGTTTTCTGGGATTAGCTCCACCTCGCGGCTTGGCAACCCTCTGTACCGACCATTGTAGCACGTGTGTAGCCCAGGCCGTAAGGGCCATGATGACTTGACGTCATCCCCACCTTCCTCCGGTTTGTCACCGGCAGTCTCCTTAGAGTGCCCACCATTACGTGCTGGTAACTAAGGACAAGGGTTGCGCTCGTTACGGGACTTAACCCAACATCTCACGACACGAGCTGACGACAGCCATGCAGCACCTGTCTCAATGTTCCCGAAGGCACCAATCCATCTCTGGAAAGTTCATTGGATGTCAAGGCCTGGTAAGGTTCTTCGCGTTGCTTCGAATTAAACCACATGCTCCACCGCTTGTGCGGGCCCCCGTCAATTCATTTGAGTTTTAACCTTGCGGCCGTACTCCCCAGGCGGTCAACTTAATGCGTTAGCTGCGCCACTAAGAGCTCAAGGCTCCCAACGGCTAGTTGACATCGTTTACGGCGTGGACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCACCTCAGTGTCAGTATCAGTCCAGGTGGTCGCCTTCGCCACTGGTGTTCCTTCCTATATCTACGCATTTCACCGCTACACAGGAAATTCCACCACCCTCTACCATACTCTAGCTCGTCAGTTTTGAATGCAGTTCCCAGGTTGAGCCCGGGGATTTCACATCCAACTTAACGAACCACCTACGCGCGCTTTACGCCCAGTAATTCCGATTAACGCTTGCACCCTCTGTATTACCGCGGCTGCTGGCACAGAGTTAGCCGGTGCTTATTCTGTCGGTAACGTCAAAACAATTACGTATTAGGTAACTGCCCTTCCTCCCAACTTAAAGTGCTTTACAATCCGAAGACCTTCTTCACACACGCGGCATGGCTGGATCAGGCTTTCGCCCATTGTCCAATATTCCCCACTGCTGCCTCCCGTAGGAGTCTGGACCGTGTCTCAGTTCCAGTGTGACTGATCATCCTCTCAGACCAGTTACGGATCGTCGCCTTGGTGAGCCATTACCTCACCAACTAGCTAATCCGACCTAGGCTCATCTGATAGCGCAAGGCCCGAAGGTCCCCTGCTTTCTCCCGTAGGACGTATGCGGTATTAGCGTCCGTTTCCGAGCGTTATCCCCCACTACCAGGCAGATTCCTAGGCATTACTCACCCGTCCGCCGCTCTCAAGAGGTGCAAGCACCTCTCTACCGCTCGACTTGCATGTGTTAGGCCTGCCGCCAGCGTTCAATCTGAGCCATGATCAAACTCTTCAGTTCAAA
>NODE_32_length_47877_cov_11.378513:46285-47829
TTTGAACTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAGAGAGGTGCTTGCACCTCTTGAGAGCGGCGGACGGGTGAGTAATGCCTAGGAATCTGCCTGGTAGTGGGGGATAACGCTCGGAAACGGACGCTAATACCGCATACGTCCTACGGGAGAAAGCAGGGGACCTTTGGGCCTTGGGCTATCAGATGAGCCTAGGTCGGATTAGCTAGTTGGTGAGGTAATGGCTCACCAAGGCGACGATCCGTAACTGGTCTGAGAGGATGATCAGTCACACTGGAACTGAGACACGGTCCAGACTCCTACGGGAGGCAGCAGTGGGGAATATTGGACAATGGGCGAAAGCCTGATCCAGCCATGCCGCGTGTGTGAAGAAGGTCTTCGGATTGTAAAGCACTTTAAGTTGGGAGGAAGGGCAGTTACCTAATACGTAATTGTTTTGACGTTACCGACAGAATAAGCACCGGCTAACTCTGTGCCAGCAGCCGCGGTAATACAGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTCGTTAAGTTGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATTCAAAACTGACGAGCTAGAGTATGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAAATGCGTAGATATAGGAAGGAACACCAGTGGCGAAGGCGACCACCTGGACTGATACTGACACTGAGGTGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCAACTAGCCGTTGGGAGCCTTGAGCTCTTAGTGGCGCAGCTAACGCATTAAGTTGACCGCCTGGGGAGTACGGCCGCAAGGTTAAAACTCAAATGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGCCTTGACATCCAATGAACTTTCCAGAGATGGATTGGTGCCTTCGGGAACATTGAGACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGTAACGAGCGCAACCCTTGTCCTTAGTTACCAGCACGTAATGGTGGGCACTCTAAGGAGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAGTCATCATGGCCCTTACGGCCTGGGCTACACACGTGCTACAATGGTCGGTACAGAGGGTTGCCAAGCCGCGAGGTGGAGCTAATCCCAGAAAACCGATCGTAGTCCGGATCGCAGTCTGCAACTCGACTGCGTGAAGTCGGAATCGCTAGTAATCGCGAATCAGAATGTCGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCATGGGAGTGGGTTGCACCAGAAGTAGCTAGTCTAACCTTCGGGAGGACGGTTACCACGGTGTGATTCATGACTGGGGTGAAGTCGTAACAAGGTAGCCGTAGGGGAACCTGCGGCTGGATCACCTCCTTAATCG
>NODE_21_length_103352_cov_10.011827:35530-37075
TTTGAACTGAAGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATGCAAGTCGAGCGGTAGAGAGGTGCTTGCACCTCTTGAGAGCGGCGGACGGGTGAGTAATGCCTAGGAATCTGCCTGGTAGTGGGGGATAACGCTCGGAAACGGACGCTAATACCGCATACGTCCTACGGGAGAAAGCAGGGGACCTTCGGGCCTTGCGCTATCAGATGAGCCTAGGTCGGATTAGCTAGTTGGTGAGGTAATGGCTCACCAAGGCGACGATCCGTAACTGGTCTGAGAGGGTGATCAGTCAGATTTGAACTGGGACATGGTCAAGACTGCTAGGGGAGGCAGCAGTGGGGAATATTGGAGAATGGGCGAAAGCCTGATCCAGCCATGCAGCGTGTGTGAAGAAGGTCTTCGGATTGTAAAGGACTTTAAGTTGGGGGGAATGGCAGTTACCTAATACGTAATTGTGTTGACGTTACCGACAGAATAAGCACCGGCTAACTCTGTGCCAGCAGCCGCGGTAATACAGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTCGTTAAGTTGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATTCAAAACTGACGAGCTAGAGTATGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAAATGCGTAGATATAGGAAGGAACACCAGTGGCGAAGGCGACCACCTGGACTGATACTGACACTGAGGTGCGAAAGCGTGGGGAGCAAACAGGATTAGATACCCTGGTAGTCCACGCCGTAAACGATGTCAACTAGCCGTTGGGAGCCTTGAGCTCTTAGTGGCGCAGCTAACGCATTAAGTTGACCGCCTGGGGAGTACGGCCGCAAGGTTAAAACTCAAATGAATTGACGGGGGCCCGCACAAGCGGTGGAGCATGTGGTTTAATTCGAAGCAACGCGAAGAACCTTACCAGGCCTTGACATCCAATGAACTTTCCAGAGATGGATTGGTGCCTTCGGGAACATTGAGACAGGTGCTGCATGGCTGTCGTCAGCTCGTGTCGTGAGATGTTGGGTTAAGTCCCGTAACGAGCGCAACCCTTGTCCTTAGTTACCAGCACGTAATGGTGGGCACTCTAAGGAGACTGCCGGTGACAAACCGGAGGAAGGTGGGGATGACGTCAAGTCATCATGGCCCTTACGGCCTGGGCTACACACGTGCTACAATGGTCGGTACAGAGGGTTGCCAAGCCGCGAGGTGGAGCTAATCCCAGAAAACCGATCGTAGTCCGGATCGCAGTCTGCAACTCGACTGCGTGAAGTCGGAATCGCTAGTAATCGCGAATCAGAATGTCGCGGTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACACCATGGGAGTGGGTTGCACCAGAAGTAGCTAGTCTAACCTTCGGGAGGACGGTTACCACGGTGTGATTCATGACTGGGGTGAAGTCGTAACAAGGTAGCCGTAGGGGAACCTGCGGCTGGATCACCTCCTTAATCGA

- open your web browser and head over to Ribosomal Database Project’s SeqMatch tool   
- upload the resulting FASTA file from the previous step.   
- This program will attempt to identify the 16S sequences as precisely as possible by comparing them to a database of high-quality, curated sequences, obtained from known bacteria.   
- Genus level identification and explanation
> Pseudomonas, from the hierarchy view of the result, all five fasta rRNA strains are computed as belong to the Pseudomonas genus, and thus it has a strong confidence to be a pseudomonas. Moreover, from the output seqmatch txt output, we can see that despite a variation between the matched species, all the matching result with high confidence are in the pseudomonas genus. 


## Genome annotation
upload genome to two remote annotation services that will perform the annotation automatically:   
1. RAST 
    1. register for an account
    2. Run fasta
2. BASys   
3. DFAST 

### RAST Result


```python
from IPython.display import Image, display
img6 = Image(filename = "RAST_table.png")
img6
```




![png](output_18_0.png)




```python
from IPython.display import Image, display
img7 = Image(filename = "RAST_graph.png")
img7
```




![png](output_19_0.png)




```python
from IPython.display import Image, display
img8 = Image(filename = "RAST_output.png")
img8
```




![png](output_20_0.png)



### BASys Result


```python
from IPython.display import Image, display
img1 = Image(filename = "BASys_output.png")
img1
```




![png](output_22_0.png)




```python
from IPython.display import Image, display
img2 = Image(filename = "BASys_summary.png")
img2
```




![png](output_23_0.png)




```python
from IPython.display import Image, display
img5 = Image(filename = "BASys_table.png")
img5
```




![png](output_24_0.png)



### DFAST run output


```python
from IPython.display import Image, display
img3 = Image(filename = "DFAST_output.png")
img3
```




![png](output_26_0.png)




```python
from IPython.display import Image, display
img4 = Image(filename = "DFAST_table.png")
img4
```




![png](output_27_0.png)



### Inspection of gene of interests
After the server run was done, a close inspection on the table of annotated features is done, and the following traits are selected for the genome analysis write-up:   
1. Antibiotic related:
    - antibiotic synthesis monooxygenase
    - antibiotic synthesis protein MbtH
    - type 2 antibiotic synthesis protein
2. Resistance related:
    - resistance-nodulation-cell division (RND) efflux membrane fusion protein
    - Bcr/CflA family drug resistance efflux transporter
    - fusaric acid resistance protein
    - multidrug resistance protein PmpM
    - azaleucine resistance protein AzlC
    - acriflavine resistance protein B
    - Glyoxalase/Bleomycin Resistance Protein/Dioxygenase
    - beta-lactamase
3. Toxin related:   
    - addiction module toxin, HicA family protein
    - toxin HipA
4. Pili related:
    - pilus assembly protein TadE
    - type IV pilus biogenesis protein PilZ
    - type IV-A pilus assembly ATPase PilB
5. Other:
    - penicillin-binding protein
    - penicillin amidase
    - cephalosporin hydroxylase
    - Biofilm PGA synthesis auxiliary protein PgaD

### Write-Up of Genome Analysis:
> From the genome annotation we gather, there are multiple type IV pilus assembly and biogenesis proteins including TadE, PilZ, and PilB. Thus, it is clear that the genus has adapted to live in aqueous environment. Moreover, as our annotation did not indicate the characteristics of ocean strains, which have much more sensors than freshwater strains, the relatives are of higher possibility to live in a freshwater or land region. 

> There's no auxotrophy related genes being annotated in all three servers' outputs, thus our strain is of high possibility to produce all the essential amino acids by itself. Similarly, CRISPR cas enzymes and short tandem repeats were not enriched out from the annotation. Despite a previous report of the type 2 CRISPR system in pseudomonas, our strain does not seem to come up with the CRISPR system from the annotation result. There are no virulence or bacteriophage related horizontal transfer being observed from the feature table. However, toxin related genes including HicA and HipA are observeed. 

> From literature, HicA, a mRNA interferase toxin, is originated in E.Coli species and so there's a horizontal transfer of this toxin gene from another species to our strain [1]. Moreover, multiple antibiotic production and resistance genes are present, suggesting that this Pseudomonas strain is capable of antibiotic production and antibiotic resistance. Specifically, from pFam MbtH,antibiotic synthesis monooxygenase and type 2 antibiotic synthesis protein are responsible for producing antibiotics. There are a number of resistance related proteins produced by the strain; resistance-nodulation-cell division (RND) efflux membrane fusion protein and Bcr/CflA family drug resistance efflux transporter is related to the Efflux-mediated multiresistance in Gram-negative bacteria, and the strain has specific genes responsible for the resistance towards various antibiotin in plants related to fusaric acid, azaleucine, acriflavine and bleomycin. Finally, the presence of beta-lactamse,penicillin amidase, and cephalosporin hydroxylase also functions as potential antibiotic-degrading enzymes, so that the strain may be resistant to penicillin and cephalosporin.

> Based on the phenotype analysis above, the bacteria is capable for binding to human tissue via pilus and form biofilm by Biofilm PGA synthesis auxiliary protein PgaD, can secrete toxic components and infect the mammalian tissue,and is also hard to treat due to the various display of resistance to antibiotics. The research also supports that biofilms of Pseudomonas can cause chronic opportunistic infections and cannot be treated effectively with traditional antibiotic therapy [2]. Thus, both genome annotation and secondary research lead to the consistent conclusion that the strain may be human pathogen.



### References:

1. Poole K (January 2004). "Efflux-mediated multiresistance in Gram-negative bacteria". Clinical Microbiology and Infection. 10 (1): 12–26. doi:10.1111/j.1469-0691.2004.00763.x. PMID 14706082.  
2. Cornelis P (2008). Pseudomonas: Genomics and Molecular Biology (1st ed.). Caister Academic Press. ISBN 1-904455-19-0.
