{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Procedural Notes  \n",
    "\n",
    "### Enzyme/Pathway Selection  \n",
    "Kegg glycolysis/Glucogenesis-Reference Pathway:  \n",
    "\n",
    ">Enzymes selected for each pathway are:  \n",
    ">- Glycolysis: K00844, K11645, K01689, K16370  \n",
    ">- Citric Acid Cycle: K01647, K00030, K01900, K00116  \n",
    ">- Pentose Phosphate Pathway: K13937, K00036, K01807, K06859  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating the genes, pathways and enzymes table with respective assigned columns\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "# to avoid conflict with previously built table\n",
    "c.execute(\"\"\"DROP TABLE genes\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE pathways\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE enzymes\"\"\")\n",
    "\n",
    "# Caution: use INTEGER instead of INT for primary key generation\n",
    "c.execute(\"\"\"CREATE TABLE genes (id INTEGER PRIMARY KEY, name TEXT, description TEXT, organism TEXT, nucleotide_sequence TEXT,\n",
    "chromosome TEXT, location TEXT, start INT, end INT, strand VARCHAR(1), translated_sequence TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathways (id INTEGER PRIMARY KEY, name TEXT, description TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzymes (id INTEGER PRIMARY KEY, kegg_id TEXT, name TEXT, function TEXT, ec_number TEXT);\"\"\")\n",
    "\n",
    "conn.commit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserting values into the pathway table\n",
    "1. get the names of the pathways\n",
    "2. search up the description of each pathway\n",
    "3. Take in two list with equal length and insert the into the corresponding field"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO pathways (name, description) VALUES(?,?)'''\n",
    "\n",
    "names = [\"Glycolysis\",\"Citric Acid Cycle\",\"Pentose Phosphate\"]\n",
    "descriptions = [\"the breakdown of glucose by enzymes, releasing energy and pyruvic acid.\" , \"the sequence of reactions by which most living cells generate energy during the process of aerobic respiration. It takes place in the mitochondria, consuming oxygen, producing carbon dioxide and water as waste products, and converting ADP to energy-rich ATP\",\"a metabolic pathway parallel to glycolysis, It generates NADPH and pentoses (5-carbon sugars) as well as ribose 5-phosphate, the last one a precursor for the synthesis of nucleotides\"]\n",
    "\n",
    "\n",
    "# Take in two list with equal length and insert the into the corresponding field\n",
    "for i in range(len(names)):\n",
    "    name = names[i]\n",
    "    description = descriptions[i]\n",
    "    temp_value = (name,description)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query the name and ec_number of the selected Enzymes from KEGG\n",
    "1. Make a sample.list file in the current directory with the following lines  \n",
    "K00844   \n",
    "K11645   \n",
    "K01689  \n",
    "K16370  \n",
    "K01647  \n",
    "K00030  \n",
    "K01900  \n",
    "K00116  \n",
    "K13937  \n",
    "K00036  \n",
    "K01807  \n",
    "K06859  \n",
    "\n",
    "2. In the terminal execute: `curl -g -s -S http://rest.kegg.jp/list/ko | grep -f sample.list | sed \"s/cpd\\://\" > sample_extracted.table.txt`\n",
    "\n",
    "3. Inspect the output file sample_extracted.table.txt  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ko:K00030\tIDH3; isocitrate dehydrogenase (NAD+) [EC:1.1.1.41]\n",
      "\n",
      "ko:K00036\tG6PD, zwf; glucose-6-phosphate 1-dehydrogenase [EC:1.1.1.49 1.1.1.363]\n",
      "\n",
      "ko:K00116\tmqo; malate dehydrogenase (quinone) [EC:1.1.5.4]\n",
      "\n",
      "ko:K00844\tHK; hexokinase [EC:2.7.1.1]\n",
      "\n",
      "ko:K01647\tCS, gltA; citrate synthase [EC:2.3.3.1]\n",
      "\n",
      "ko:K01689\tENO, eno; enolase [EC:4.2.1.11]\n",
      "\n",
      "ko:K01807\trpiA; ribose 5-phosphate isomerase A [EC:5.3.1.6]\n",
      "\n",
      "ko:K01900\tLSC2; succinyl-CoA synthetase beta subunit [EC:6.2.1.4 6.2.1.5]\n",
      "\n",
      "ko:K06859\tpgi1; glucose-6-phosphate isomerase, archaeal [EC:5.3.1.9]\n",
      "\n",
      "ko:K11645\tfbaB; fructose-bisphosphate aldolase, class I [EC:4.1.2.13]\n",
      "\n",
      "ko:K13937\tH6PD; hexose-6-phosphate dehydrogenase [EC:1.1.1.47 3.1.1.31]\n",
      "\n",
      "ko:K16370\tpfkB; 6-phosphofructokinase 2 [EC:2.7.1.11]\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['ko:K00030',\n",
       "  'ko:K00036',\n",
       "  'ko:K00116',\n",
       "  'ko:K00844',\n",
       "  'ko:K01647',\n",
       "  'ko:K01689',\n",
       "  'ko:K01807',\n",
       "  'ko:K01900',\n",
       "  'ko:K06859',\n",
       "  'ko:K11645',\n",
       "  'ko:K13937',\n",
       "  'ko:K16370'],\n",
       " ['IDH3',\n",
       "  'G6PD,zwf',\n",
       "  'mqo',\n",
       "  'HK',\n",
       "  'CS,gltA',\n",
       "  'ENO,eno',\n",
       "  'rpiA',\n",
       "  'LSC2',\n",
       "  'pgi1',\n",
       "  'fbaB',\n",
       "  'H6PD',\n",
       "  'pfkB'],\n",
       " ['isocitrate dehydrogenase (NAD+)',\n",
       "  'glucose-6-phosphate 1-dehydrogenase',\n",
       "  'malate dehydrogenase (quinone)',\n",
       "  'hexokinase',\n",
       "  'citrate synthase',\n",
       "  'enolase',\n",
       "  'ribose 5-phosphate isomerase A',\n",
       "  'succinyl-CoA synthetase beta subunit',\n",
       "  'glucose-6-phosphate isomerase, archaeal',\n",
       "  'fructose-bisphosphate aldolase, class I',\n",
       "  'hexose-6-phosphate dehydrogenase',\n",
       "  '6-phosphofructokinase 2'],\n",
       " ['EC:1.1.1.41]',\n",
       "  'EC:1.1.1.49 1.1.1.363]',\n",
       "  'EC:1.1.5.4]',\n",
       "  'EC:2.7.1.1]',\n",
       "  'EC:2.3.3.1]',\n",
       "  'EC:4.2.1.11]',\n",
       "  'EC:5.3.1.6]',\n",
       "  'EC:6.2.1.4 6.2.1.5]',\n",
       "  'EC:5.3.1.9]',\n",
       "  'EC:4.1.2.13]',\n",
       "  'EC:1.1.1.47 3.1.1.31]',\n",
       "  'EC:2.7.1.11]'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parse in the sample_extracted.table.txt for enzymes table insertion\n",
    "def parse_extracted_table():\n",
    "    filename = \"sample_extracted.table.txt\"\n",
    "    in_file = open(filename, \"rt\")\n",
    "    lines = []\n",
    "    # separate each line in the for for inspection and downstream analysis\n",
    "    for line in in_file:\n",
    "        lines.append(line.rstrip('\\n'))\n",
    "        print(line)\n",
    "    return lines\n",
    "\n",
    "lines = parse_extracted_table()\n",
    "\n",
    "\n",
    "# the get balues function split each line in the sample_extracted.table.txt to render the kegg_id, \n",
    "# enzyme name, function, and ec_number as separate lists\n",
    "def get_values(lines):\n",
    "    kegg_ids = []\n",
    "    names = []\n",
    "    functions = []\n",
    "    ec_numbers = []\n",
    "    for line in lines:\n",
    "        # first, separate out the ec_number by the \"[\" symbol\n",
    "        line_ec = line.split(\"[\")\n",
    "        ec_number = line_ec[-1]\n",
    "        # separate the by the \";\" symbol, the first half will be kegg_id and the name, the second part function\n",
    "        line_2_part = line_ec[0].split(\";\")\n",
    "        # line_part1 contains kegg_id and the name\n",
    "        line_part1 = line_2_part[0].split()\n",
    "        # line_part2 contains function description\n",
    "        line_part2 = line_2_part[1].split()\n",
    "#         print(line_part1)\n",
    "#         print(line_part2)\n",
    "        kegg_id = line_part1[0]\n",
    "        # join the 1:end of part1 to form the name\n",
    "        temp_name = line_part1[1:]\n",
    "        temp_name[:] = [''.join(temp_name[:])]\n",
    "        name = temp_name[0]\n",
    "        # join all segments in part2 to form the function string\n",
    "        temp_function = line_part2[:]\n",
    "        temp_function[:] = [\" \".join(temp_function[:])]\n",
    "        function = temp_function[0]     \n",
    "        kegg_ids.append(kegg_id)\n",
    "        names.append(name)\n",
    "        functions.append(function)\n",
    "        ec_numbers.append(ec_number)\n",
    "    return kegg_ids,names,functions,ec_numbers\n",
    "\n",
    "\n",
    "# inspect the 3 output lists\n",
    "get_values(lines)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert values into the enzymes table with 4 equal length list names, functions, and kegg_ids and ec_numbers \n",
    "# into the corresponding columns of the enzymes table\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO enzymes (kegg_id, name, function, ec_number) VALUES(?,?,?,?)'''\n",
    "\n",
    "# get the four lists\n",
    "kegg_ids,names,functions,ec_numbers = get_values(lines)\n",
    "# insertion\n",
    "for i in range(len(names)):\n",
    "    name = names[i]\n",
    "    function = functions[i]\n",
    "    kegg_id = kegg_ids[i]\n",
    "    ec_number = ec_numbers[i]\n",
    "    temp_value = (kegg_id, name, function, ec_number)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Drosophila[ORGN] IDH3', 'Drosophila[ORGN] G6PD,zwf', 'Drosophila[ORGN] mqo', 'Drosophila[ORGN] HK', 'Drosophila[ORGN] CS,gltA', 'Drosophila[ORGN] ENO,eno', 'Drosophila[ORGN] rpiA', 'Drosophila[ORGN] LSC2', 'Drosophila[ORGN] pgi1', 'Drosophila[ORGN] fbaB', 'Drosophila[ORGN] H6PD', 'Drosophila[ORGN] pfkB', 'Escherichia coli[ORGN] IDH3', 'Escherichia coli[ORGN] G6PD,zwf', 'Escherichia coli[ORGN] mqo', 'Escherichia coli[ORGN] HK', 'Escherichia coli[ORGN] CS,gltA', 'Escherichia coli[ORGN] ENO,eno', 'Escherichia coli[ORGN] rpiA', 'Escherichia coli[ORGN] LSC2', 'Escherichia coli[ORGN] pgi1', 'Escherichia coli[ORGN] fbaB', 'Escherichia coli[ORGN] H6PD', 'Escherichia coli[ORGN] pfkB', 'Homo sapiens[ORGN] IDH3', 'Homo sapiens[ORGN] G6PD,zwf', 'Homo sapiens[ORGN] mqo', 'Homo sapiens[ORGN] HK', 'Homo sapiens[ORGN] CS,gltA', 'Homo sapiens[ORGN] ENO,eno', 'Homo sapiens[ORGN] rpiA', 'Homo sapiens[ORGN] LSC2', 'Homo sapiens[ORGN] pgi1', 'Homo sapiens[ORGN] fbaB', 'Homo sapiens[ORGN] H6PD', 'Homo sapiens[ORGN] pfkB']\n",
      "['IDH3', 'G6PD,zwf', 'mqo', 'HK', 'CS,gltA', 'ENO,eno', 'rpiA', 'LSC2', 'pgi1', 'fbaB', 'H6PD', 'pfkB', 'IDH3', 'G6PD,zwf', 'mqo', 'HK', 'CS,gltA', 'ENO,eno', 'rpiA', 'LSC2', 'pgi1', 'fbaB', 'H6PD', 'pfkB', 'IDH3', 'G6PD,zwf', 'mqo', 'HK', 'CS,gltA', 'ENO,eno', 'rpiA', 'LSC2', 'pgi1', 'fbaB', 'H6PD', 'pfkB']\n",
      "['Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Drosophila', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Escherichia coli', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens', 'Homo sapiens']\n",
      "['isocitrate dehydrogenase (NAD+)', 'glucose-6-phosphate 1-dehydrogenase', 'malate dehydrogenase (quinone)', 'hexokinase', 'citrate synthase', 'enolase', 'ribose 5-phosphate isomerase A', 'succinyl-CoA synthetase beta subunit', 'glucose-6-phosphate isomerase, archaeal', 'fructose-bisphosphate aldolase, class I', 'hexose-6-phosphate dehydrogenase', '6-phosphofructokinase 2', 'isocitrate dehydrogenase (NAD+)', 'glucose-6-phosphate 1-dehydrogenase', 'malate dehydrogenase (quinone)', 'hexokinase', 'citrate synthase', 'enolase', 'ribose 5-phosphate isomerase A', 'succinyl-CoA synthetase beta subunit', 'glucose-6-phosphate isomerase, archaeal', 'fructose-bisphosphate aldolase, class I', 'hexose-6-phosphate dehydrogenase', '6-phosphofructokinase 2', 'isocitrate dehydrogenase (NAD+)', 'glucose-6-phosphate 1-dehydrogenase', 'malate dehydrogenase (quinone)', 'hexokinase', 'citrate synthase', 'enolase', 'ribose 5-phosphate isomerase A', 'succinyl-CoA synthetase beta subunit', 'glucose-6-phosphate isomerase, archaeal', 'fructose-bisphosphate aldolase, class I', 'hexose-6-phosphate dehydrogenase', '6-phosphofructokinase 2']\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "#Entrez.esearch(db,term, sort)\n",
    "#Entrez.efetch(db, id, rettype, retmode)\n",
    "#for enzymes and genes only; Entrez doesn't have pathway info\n",
    "\n",
    "\n",
    "# a helper function to search and fetch for gene, nucleotide, and protein by specifying the db_type and the term\n",
    "def search_fetch(db_type,term):\n",
    "    Entrez.email = 'ych323@berkeley.edu'\n",
    "    handle = Entrez.esearch(db = db_type,\n",
    "                       term = term,\n",
    "                       sort= 'relevance',\n",
    "                       idtype= 'acc')\n",
    "    fetched_dict = Entrez.read(handle)\n",
    "#     print(fetched_dict)\n",
    "    if not fetched_dict[\"IdList\"]:\n",
    "        data = \"None\"\n",
    "        return data\n",
    "    handle = Entrez.efetch(db = db_type, id = fetched_dict[\"IdList\"][0], rettype = 'fasta', retmode = 'text')\n",
    "    data = handle.read()\n",
    "    return data\n",
    "\n",
    "\n",
    "# build a term list of all the search \"terms species[ORGN] enzyme_name\"\n",
    "# build the 3 * 12 entries of name, organism, and description of the gene\n",
    "species_list = [\"Drosophila\",\"Escherichia coli\",\"Homo sapiens\"]\n",
    "ter_list = [str(species) + \"[ORGN]\" + \" \" for species in species_list ]\n",
    "term_list = [str(term) + str(name) for term in ter_list for name in names]\n",
    "name_list = [str(name) for term in ter_list for name in names]\n",
    "organism_list = [str(species) for species in species_list for name in names]\n",
    "descriptions_list = [str(function) for species in species_list for function in functions]\n",
    "\n",
    "# for the strand column, the DNA nucleotide sequence stored in GeneBank are all 5' to 3' of the sense strand, and so the \n",
    "# sign of the strand should all be \"+\"\n",
    "strand_list = ['+']*len(term_list)\n",
    "\n",
    "# inspect the built lists \n",
    "print(term_list)\n",
    "print(name_list)\n",
    "print(organism_list)\n",
    "print(descriptions_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Drosophila[ORGN] IDH3\n",
      "Drosophila[ORGN] G6PD,zwf\n",
      "parsed\n",
      "Drosophila[ORGN] mqo\n",
      "parsed\n",
      "Drosophila[ORGN] HK\n",
      "parsed\n",
      "Drosophila[ORGN] CS,gltA\n",
      "parsed\n",
      "Drosophila[ORGN] ENO,eno\n",
      "parsed\n",
      "Drosophila[ORGN] rpiA\n",
      "Drosophila[ORGN] LSC2\n",
      "Drosophila[ORGN] pgi1\n",
      "Drosophila[ORGN] fbaB\n",
      "Drosophila[ORGN] H6PD\n",
      "Drosophila[ORGN] pfkB\n",
      "parsed\n",
      "Escherichia coli[ORGN] IDH3\n",
      "Escherichia coli[ORGN] G6PD,zwf\n",
      "parsed\n",
      "Escherichia coli[ORGN] mqo\n",
      "parsed\n",
      "Escherichia coli[ORGN] HK\n",
      "parsed\n",
      "Escherichia coli[ORGN] CS,gltA\n",
      "parsed\n",
      "Escherichia coli[ORGN] ENO,eno\n",
      "parsed\n",
      "Escherichia coli[ORGN] rpiA\n",
      "parsed\n",
      "Escherichia coli[ORGN] LSC2\n",
      "Escherichia coli[ORGN] pgi1\n",
      "Escherichia coli[ORGN] fbaB\n",
      "parsed\n",
      "Escherichia coli[ORGN] H6PD\n",
      "Escherichia coli[ORGN] pfkB\n",
      "parsed\n",
      "Homo sapiens[ORGN] IDH3\n",
      "parsed\n",
      "Homo sapiens[ORGN] G6PD,zwf\n",
      "parsed\n",
      "Homo sapiens[ORGN] mqo\n",
      "Homo sapiens[ORGN] HK\n",
      "parsed\n",
      "Homo sapiens[ORGN] CS,gltA\n",
      "parsed\n",
      "Homo sapiens[ORGN] ENO,eno\n",
      "parsed\n",
      "Homo sapiens[ORGN] rpiA\n",
      "parsed\n",
      "Homo sapiens[ORGN] LSC2\n",
      "parsed\n",
      "Homo sapiens[ORGN] pgi1\n",
      "Homo sapiens[ORGN] fbaB\n",
      "parsed\n",
      "Homo sapiens[ORGN] H6PD\n",
      "parsed\n",
      "Homo sapiens[ORGN] pfkB\n",
      "parsed\n",
      "['GCGATCGTCTCAAGCTACATATCGTTAATATTTTAGGGCTCGCTGAAAACCAAAAGGCACCGAAAACTAAACCTTCTTGCCAAACTACTAAGCTTGCGCACCAAAATAAAATAAATAAAGGGGAAAAGACGATTAAACAAAAACTTTGTTCACGTTGCGCTTGCGTGCAAGTGCAGCCGAGTGTGTGTGTTTGTGTGTGTGTATGCCTGTGTGAGTGTGCGTGTGTTTTGCCAACCGCAGCAAAAATCACAATATTTTCTGCGCTTGACCTTTATTCAAATTATTGGTCAACAGTTTTATTAATTGCTGCACGGAAATCATTAGAAAAGTTGGCTTCAAAAATTGACATCAAAGAAGACTGTCAGTGAGAAATAAGTAAACAAATAACAGTTAGCACACACACACTACGAAAAATAAACATGGAGTGCAAATGCTCAAAGCCCTCGGGCACCTTGGATCGCATCATCAACACTCTGAAGTGCTCAACTATGATATGCGAGGGCACACACTTTGACGGTCAAATTCCACACACTTTTGTGGTGTTTGGTGCATCGGGCGATCTGGCTAAGAAGAAGATATATCCAACACTATGGTGGCTATATCGCGACGATCTGCTGCCGAAGCCAACAAAAATTTGCGGCTATGCGCGCTCCAAGATGACAGTGGACCAGTTGAAGAATCTATGTGAGCAGTATATGAAGGTGCAACCGCATGAACAGATGAAATATAATGAATTCTGGCGCTTGAATAACTATGTTTCCGGACCATACGATGACGCCCAGGGCTACCAGGCGCTCAACCAGCACATGGAGTCGATCGAGAATCGGCTTAATGCGAATCGCATCTTCTATTTGGCACTGCCGCCCACAGTCTACAACCAGGTGACGTTCAACATCAAGCACAACTGCATGTCCAAGGGCGGCTGGAATCGCGTGATCGTCGAGAAGCCATTCGGGCGAGATGATGTGACCTCGAAGGCGCTGAGCGATCATCTGGCCGGCCTGTTCGAGGAGGAGCAACTGTATCGCATCGATCACTACTTGGGCAAGGAGATGGTGCAAAACTTGATGACCATACGGTTCGGCAACAAGATTCTGAGCTCAACATGGAATCGCGAGAATATTGCCTGCGTCCTGATCACGTTCAAGGAGCCGTTCGGTACGCAGGGGCGTGGCGGCTACTTTGACGAGTTCGGCATCATACGAGATGTGATGCAGAATCATCTGCTGCAGATTCTATCGCTGGTGGCCATGGAGAAGCCATGCAGCTGCCATCCGGACGACATACGTGATGAGAAGGTCAAGGTGCTGAAGTGCATCAAGCCATTGGAGCTGAGCGACATGGTCCTGGGCCAGTATGTGGGCAATCCCGCAGGCAGCACGGCTGAGGAGCAGTGCGGCTACCTGGATGATCCCACCGTGAGCAAGACATCGAACACGCCCACCTATGCCATGGCCGTCATTCAGATCAACAACGAGCGCTGGCAGGGCGTGCCGTTCATACTGCGCTGCGGCAAGGCATTGAACGAGCGCAAGGCCGAGGTCCGCATCCAATATCAGGATGTGCCCGGCGACATCTTCGAGGGCAACTCGAAGCGCAACGAGCTGGTGATACGAGTGCAGCCGGGCGAGGCGCTCTACTTCAAGATGATGACCAAGAGTCCGGGCATTACGTTCGACATTGAGGAGACCGAACTGGATCTCACCTATGAGCATCGCTATAAACATTCCGTGCTGCCGGACGCCTACGAGCGCCTCATTCTGGACGTCTTCTGCGGCTCCCAGATGCATTTCGTGCGCTCCGACGAGCTGCGCGAGGCGTGGCGCATCTTCACGCCCATTCTGCACAAGATCGAGCATGAGCGCATCCCACCCATTCCCTATCCGTACGGCTCCCGCGGCCCCACTGAGGCCGATCGCAAGTGCGTCGAGAATAACTTCATCTACTCGGCGTCCTACAAATGGCATGGCAATAAGGCGGATGCATCTCATCTTTGATTTGGGACTCCATGGGATCCATTGGCTTTTTGGTGTTTATGTTGATCATTAACTAACTCTAGCGACATTTATACATATACTCTGCATATATATCTATGTATATATATGCATATAATAGGCGTT', 'ATGAGCGCCACGCTGGGGGCTTTTATCAGACAGCTTCAGCCCGATTGGAGCATCAGCATTTTTGAACGCCTGGATAATGTGGCGGAAGAAAGTTCGGATGCCTGGAACAACGCGGGAACAGGCCATTCCGCGTTATGTGAATTGAATTACACCCCCCAACTTGCTGACGGCAGTGTAGATATCAGCAAGGCTGTGGGTGTGAATGAATCGTTCCAGATTTCACGCCAGTTCTGGGCTTATCTGGTGGAGCAGGGGATTATTGCCAACCCGCGCGACTTCATTACGCCGGTGCCGCATATGAGCTTTGTGTGGGGGGAGCAGAATGTCTCCTTTCTGCGGAAGAGGTATGAGGCGTTAAGCGCTCATCCATTGTTTGCCGGGATGGAGTATTCGGAAGACCCGGCCCAGATGGCCCAGTGGATGCCATTGGTCATGCAAAATCGGCCGGCCGGTCAGAAGCTTGCCATTACCCGTAGCCTGAATGGAACGGACGTAAATTTTGGGGCATTAACGCGTCTACTGTTTGGCTATTTGGCAAGCACACCCGCTTGCACCTTGCATACGAAACATGATGTTCAGGACATTACCCGAGGAAGTGACGGGCGTTGGCGCGTACAGGTGCAGGACCTACGCCTGAACACACAACGCACCGTCAGTGCAAAATTTGTGTTTATTGGCGGCGGCGGCGGTGCTCTGCCACTGTTGCAGAAAACAGGAATCCCTGAATCACGAGGTGTGGGAGGTTTCCCTGTCAGTGGCCAGTTCCTACGGTGCAAAAACCCGGACATCATCGCGCAGCACCATGCAAAAGTATATGGCAAGGCGTCAGTAGGGGCTCCCCCCATGTCCGTGCCACATCTGGATACTCGCGTAATTAACGGGCAACGGGCGTTGCTTTTCGGACCATACGCCGGATTTTCGACGCGCTTCCTGAAACACGGCTCCCTAATGGACCTGCCAAAATCTGTCAGAGGCAATAATCTTGGTGCCATGTTGGCTGTGGCACGAGATAACTGGGCGCTGACAAAATATCTCATTCAGCAGGTCATGCAGTCCAACAATGACCGCATGAATGCTTTGCGTGACTTTGTGCCCGATGCTCGGCCGGAAGACTGGGAACTGGTCACGGCTGGCCAGCGTGTACAAATTATCAAGAAGGATGCCAAAAAAGGAGGCGTTCTTCAGTTTGGGACAGAAGTTATTGCCAGCCAGGACGGATCGGTTGCGGCTCTGCTGGGCGCGTCTCCGGGGGCATCAACCGCTGCCCCAATTATGCTGACAGTTCTGAAAAAATGCTTTGCTGACAAGCTGCCAGAATGGGAGGCAAAGCTGAAAATAATGGTGCCGTCTTATGGGCAGACATTGGCAAACAACCCAGAACTCTGCACGCAACTGCGAGACAAGACGACGGACGTTCTGAAACTGAAAGACTGA', 'CAGAATAAATATCTTCAATAGCGACAAACAAATAAACATTTATTGGACTGGACTTTGCCTTAAAAGTATGCTTTTTAGAAAACAAAATCCCATGTGATTTGTAGTCCAGTCTAAGTAGTAAGATACAACTAAAATCAACCAAAACAAATGCAGAAGTGAAATGGAAATATCTGTACTTGAAGCAAACAAAAAAATTTCTGTGGTTATATAAAATACTTTGAAATTATGATAAACTTTCAAGTAATCATATCATTGAAAATAGTAAATATGCGAAAGCAAAAACAAACTCAAGTAACCTAAAAGATGAAGGTATAATTCTAGAGAAACTACAAAAACTAAAAAATTTGAAAGCTAATAGGTGTTGAAAATACTACGTACAACTTAACAAAACTATAAACTCTTAACTACAAAAATAAATATTTATATTTAACAAGTAGCTGAATCAATAGCAGACATTTTAAATGTCAAGGCAAACTTCAATTAGAAATGCAAAAGAAAACCACACAAATATAAATAAAACCTTGCAGCTATTAAAATTGAAATCAAATGCCTAAATAAATAATAGAAGAAATTCCAATAAACACGCAGTAGCTAAATAAATATGAAGTGTGCAAACCTAGGATAAACAATGGCGTCCAATAGTATCTAATAAAAATATAAATAGAAAACGTAGGCAAGCAAAATAACAACACAAGTGAAAGTAAACTTTGATAAATATTAAAATTTAGTTAAATCATAACAAAAGCAGTCGAACTATAAGCACAATGTCTATGGCATTATGCAATTTAAATGGTGACGGCTCGGCGGCGCAGTCGACGTCGCAGTCGCAGTCGCCGGCCGCAACCGCCGCAGCAGCTCCGTTGCTGTCTCACTCGCACTCGCACACGCTACAGCAGGAGAGTACGCCGCTGCTCTTGGGACACGAGCAGAGCGGCAGCGCTGCTGCAGAGGGCAGCGGCGGCGGTGACGTCACTGACGGAGCGGTAACTTCCATTGAAATGCCCACTGTTGTTGCCGACGATACCGGTGTACCGCCGCTGCCTCTGCCTGTGCCGCAGTCGACGCCGCAGCCGCTGCTTATGGCTCTGCCCGCCAATTTAAACTTCATAACGGGACCCACGACTCAGCAACTGATGATCGGTAATGGCGCCATGGGCGTAATTCCCTCGACCAATGACAGCAACAATAACAACAACAACAATGTGGACAATACTAGCGACGAATCGAACCCGGTCACCATTTACAGGTGCCGGGCTCCCATTGCGTCACTCGACTGTATGGAGGAGTTCAGTGGTCGTTCGATCAGTTTGGGCTCCAATCCGGCGTTACCACTGCGCCATGGATCCACACCGACGCCGGGGCTGCGGTACAAGAATCTGGGCAAGAGCGGGCTGCGTATCTCGAATGTGGGATTGGGTACCTGGCCGGTCTTCTCGCCAGGCGTGAGCGATGACCAGGCGGAGGCAATCCTTAAGCTGGCCATCGAGAGCGGAATCAATCTGTTCGACATCTCGGAGGCGCACTCGGAGACGGAGATCGGCAAGATACTGCAGCGGGCGGGCTGGAAGAGGACCGCCTATGTCATCACCACGAAGGTCTACTGGAGCACCAAGTCGGAGGAACGGGGCCTCTCCCGGAAACACATCATCGAATGTGTTCGAGCCAGTTTACAGCGATTGCAGCTGCAGTACATCGATATCGTCATCATCCACAAGGCAGATCCCATGTGTCCCATGGAGGTGGTGCGCGCCATGAGCTACGTGATACAGCAGGGCTGGGCAATGTACTGGGGCACCGCTCGATGGAGCCAGGTGGAGATCATGGAGGCCTACACCAACTGCCGCCAGTTCAACTGCATCACGCCCATTGTGGAGCAGTCCGAGTACCACATGTTCTGCCGCGAAAAGTGCGAGCTCTACCTGCCGGAGATGTACAACAAGATCGGAGTGGGCCTCATGGCCTGGGGTCCACTCTCAATGGCCCTCAGCGACACCCAGAATGGGGACAAGCTTTTCCTGCCCAAGGGCTCCTTCAAGACGAAGAGCTTCTCCTGGACCGAGGACGAGATCAATCGTAATGCCGCTCTGTCGCCGCAGGGCAGTTGGGGTAAGGACCGGATCGATGAGGGTCGTCGCCACTGCGACCGCCTCCGCGACCTTGCCGCCCTCGCCGAGAAGCTGGGCTGCAGCCCCACCCAATTGTCGATCGCCTGGTCGCTGAAACATGAGCCAGTGCAATGCCTGTTGCTGGGCGCCACATCGGCGGAGCAGCTGCACCAGAGTCTGCAGTCGTTGCAGCTGCTGCCACGACTCTCCTCGAGCGTTATGCTGGAGCTGGAAAGGATATTGGAAAACAAGCCGGTGCGGCCGCCGATGATCTCGACCTTGGCGCTCCGGTGACAATCAGCGTGCCCGCAGGGGCCGCTCAGCCTGAGCGGAGGCGGACCCTGCGGCGCCGATTCGCCCAAGCACGAGGATGAGGCCTTCATCGAGGAGGTGGACGCTGCCAAGGACGCTGCCAAGCAAGGCTTCTGCCTCATCCAGTGACAAAAGGAGTCCTTAGATAACGATAATAAATTTTTGAAAGTGATTAGCAGTAAAACTAGCCTAAGTAGCAATACCATGACGACAACAATAAGCAAGCAGGAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAGCAACAGCAGCAGCAACATCTCCAGCCTCAGCAGTATGTGCAACTACAGCAGCAACATTTGGTGGCCAGCGGCAGCACCAGCGATTCGGCAGCTGTTCCCATCGCAGCTGAACTCCAGTTCGACTCCCTGAATAACAATAGCAACAACAACAATAATAATACCAATGGCAGTCCGGAACAGGCACTGCCCTCCAGCGATGATCGCCAGAAACGTCGCTCCTTGCTCAACTTCAAGTCCCTGGACTTTCATCTAAAGTCCCTGTACAGTGGCCTACGGAGTAACTCAGGCACAACTCCACCCAGCCAACCGAATAGCTCCGTTTCGGCGGTGGATCAACCGCAGCGTAGGATGCCCTACTTGCGGGTGGAGAGTGTGGATCCCGAGGAGCCGAGAGGATCAGTTCTGCCAGGACATTCGCCCAGCTTTCTGTCACCCTACAGTGGCGTTGGCGGTGGAGCCGGAAACCTGCAGTTGCCCTCAACCGATGGAGGCACCATCAGGAGATCCTCAACATCTGACATTGTTAATTGTAGGCGGGGAGGGGGAGCAGGGGGATCGGGCGGAGGAGCAGCCTCCGCCCAGGACAGCAGGAGGCCGAGCACATCGGATTTGCTCAGGAAGGCAAGGGAACGGAAGGGCAGTGAGGCCAGGATGGGCAGAAGTGCCTCCCACAGTGGGCTAACCAGGGGCGGACCCGGCGGAGGAGGTGGTGGCCTGGGCGGAAGACGAACCAGCATGGCCTTCTAAAAAAAAGAAAATAATAGTGAAAAAAAACTATGGAAAAAAAAACTTAAGCAATGAGGGATGGCGAAAAAAACTGTTGATCAAATCTGAATCAGAAATAGGCTGACAAAATGGCGCTTATGAATTTTTATCTAGTGAAAAGAAAATTTAAAGAAAAAACAAGAAAAAAATAACAATAATAGAGCAGCTATTAAGGAACGACTTTTTTGTGGATCTATTATCATTATGATTATTATATGGTTACTACAATTGCGCCTCGAAAGAACTAAAGTTAGTAGATAAATTAAAGCAAGTAATGAACAGAAA']\n"
     ]
    }
   ],
   "source": [
    "# Get the nucleotide sequences of each enzyme through searching entrez and fetch the most relevent result\n",
    "def get_nucleotide_sequence(term_list):\n",
    "    nucleotide_list = []\n",
    "    for term in term_list:\n",
    "        print(term)\n",
    "        data = search_fetch(\"nucleotide\",term)\n",
    "        # If data is \"None\", then the entrez search has fetched no id and the idlist is empty, and \n",
    "        # so cannot be parsed by SeqIO, so should be directly append to the nucleotide list\n",
    "        if data == \"None\":\n",
    "            nucleotide_list.append(data)\n",
    "            continue\n",
    "            \n",
    "        # If data is not \"None\", then the search_and_fetch function return the most relevent sequence, wrtie data into a\n",
    "        # fasta file and then use SeqIO.parse to get the sequence to a string for sqlite insertion\n",
    "        f = open('temp.fa','w')\n",
    "        f.write(data)\n",
    "        f.close()\n",
    "        print(\"parsed\")\n",
    "        for seq in SeqIO.parse(\"temp.fa\", \"fasta\"):\n",
    "            nucleotide_list.append(str(seq.seq))\n",
    "    return nucleotide_list\n",
    "\n",
    "nucleotide_list = get_nucleotide_sequence(term_list)\n",
    "# inspect first three entries of the nucleotide_list\n",
    "print(nucleotide_list[1:4])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Drosophila[ORGN] IDH3\n",
      "Drosophila[ORGN] G6PD,zwf\n",
      "parsed\n",
      "Drosophila[ORGN] mqo\n",
      "parsed\n",
      "Drosophila[ORGN] HK\n",
      "parsed\n",
      "Drosophila[ORGN] CS,gltA\n",
      "parsed\n",
      "Drosophila[ORGN] ENO,eno\n",
      "parsed\n",
      "Drosophila[ORGN] rpiA\n",
      "Drosophila[ORGN] LSC2\n",
      "Drosophila[ORGN] pgi1\n",
      "Drosophila[ORGN] fbaB\n",
      "Drosophila[ORGN] H6PD\n",
      "Drosophila[ORGN] pfkB\n",
      "parsed\n",
      "Escherichia coli[ORGN] IDH3\n",
      "Escherichia coli[ORGN] G6PD,zwf\n",
      "parsed\n",
      "Escherichia coli[ORGN] mqo\n",
      "parsed\n",
      "Escherichia coli[ORGN] HK\n",
      "parsed\n",
      "Escherichia coli[ORGN] CS,gltA\n",
      "parsed\n",
      "Escherichia coli[ORGN] ENO,eno\n",
      "parsed\n",
      "Escherichia coli[ORGN] rpiA\n",
      "parsed\n",
      "Escherichia coli[ORGN] LSC2\n",
      "Escherichia coli[ORGN] pgi1\n",
      "Escherichia coli[ORGN] fbaB\n",
      "parsed\n",
      "Escherichia coli[ORGN] H6PD\n",
      "Escherichia coli[ORGN] pfkB\n",
      "parsed\n",
      "Homo sapiens[ORGN] IDH3\n",
      "parsed\n",
      "Homo sapiens[ORGN] G6PD,zwf\n",
      "parsed\n",
      "Homo sapiens[ORGN] mqo\n",
      "parsed\n",
      "Homo sapiens[ORGN] HK\n",
      "parsed\n",
      "Homo sapiens[ORGN] CS,gltA\n",
      "parsed\n",
      "Homo sapiens[ORGN] ENO,eno\n",
      "parsed\n",
      "Homo sapiens[ORGN] rpiA\n",
      "parsed\n",
      "Homo sapiens[ORGN] LSC2\n",
      "Homo sapiens[ORGN] pgi1\n",
      "Homo sapiens[ORGN] fbaB\n",
      "parsed\n",
      "Homo sapiens[ORGN] H6PD\n",
      "parsed\n",
      "Homo sapiens[ORGN] pfkB\n",
      "parsed\n",
      "['MECKCSKPSGTLDRIINTLKCSTMICEGTHFDGQIPHTFVVFGASGDLAKKKIYPTLWWLYRDDLLPKPTKICGYARSKMTVDQLKNLCEQYMKVQPHEQMKYNEFWRLNNYVSGPYDDAQGYQALNQHMESIENRLNANRIFYLALPPTVYNQVTFNIKHNCMSKGGWNRVIVEKPFGRDDVTSKALSDHLAGLFEEEQLYRIDHYLGKEMVQNLMTIRFGNKILSSTWNRENIACVLITFKEPFGTQGRGGYFDEFGIIRDVMQNHLLQILSLVAMEKPCSCHPDDIRDEKVKVLKCIKPLELSDMVLGQYVGNPAGSTAEEQCGYLDDPTVSKTSNTPTYAMAVIQINNERWQGVPFILRCGKALNERKAEVRIQYQDVPGDIFEGNSKRNELVIRVQPGEALYFKMMTKSPGITFDIEETELDLTYEHRYKHSVLPDAYERLILDVFCGSQMHFVRSDELREAWRIFTPILHKIEHERIPPIPYPYGSRGPTEADRKCVENNFIYSASYKWHGNKADASHL', 'MSATLGAFIRQLQPDWSISIFERLDNVAEESSDAWNNAGTGHSALCELNYTPQLADGSVDISKAVGVNESFQISRQFWAYLVEQGIIANPRDFITPVPHMSFVWGEQNVSFLRKRYEALSAHPLFAGMEYSEDPAQMAQWMPLVMQNRPAGQKLAITRSLNGTDVNFGALTRLLFGYLASTPACTLHTKHDVQDITRGSDGRWRVQVQDLRLNTQRTVSAKFVFIGGGGGALPLLQKTGIPESRGVGGFPVSGQFLRCKNPDIIAQHHAKVYGKASVGAPPMSVPHLDTRVINGQRALLFGPYAGFSTRFLKHGSLMDLPKSVRGNNLGAMLAVARDNWALTKYLIQQVMQSNNDRMNALRDFVPDARPEDWELVTAGQRVQIIKKDAKKGGVLQFGTEVIASQDGSVAALLGASPGASTAAPIMLTVLKKCFADKLPEWEAKLKIMVPSYGQTLANNPELCTQLRDKTTDVLKLKD']\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "# following similar approach as get_nucleotide to get the protein sequence from entrez search and fetch\n",
    "def get_protein(term_list):\n",
    "    protein_list = []\n",
    "    for term in term_list:\n",
    "        print(term)\n",
    "        data = search_fetch(\"protein\",term)\n",
    "        if data == \"None\":\n",
    "            protein_list.append(data)\n",
    "            continue\n",
    "        f = open('temp.fa','w')\n",
    "        f.write(data)\n",
    "        f.close()\n",
    "        print(\"parsed\")\n",
    "        for seq in SeqIO.parse(\"temp.fa\", \"fasta\"):\n",
    "            protein_list.append(str(seq.seq))\n",
    "    return protein_list\n",
    "protein_list = get_protein(term_list)\n",
    "\n",
    "#inspect the output protein list\n",
    "print(protein_list[1:3])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['None', 'Chromosome: Unknown', 'Chromosome: Un', 'Chromosome: X; Location: 1-31 cM', 'Chromosome: 3R; Location: 3-51 cM', 'Chromosome: 2L; Location: 2-3 cM', 'None', 'None', 'None', 'None', 'None', 'Chromosome: Unknown', 'None', 'None', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'None', 'Chromosome: X; Location: Xq28', 'None', 'Chromosome: 10; Location: 10q22.1', 'None', 'Chromosome: 1; Location: 1p36.23', 'Chromosome: 2; Location: 2p11.2', 'None', 'Chromosome: 12; Location: 12q24.31', 'Chromosome: 7; Location: 7p22.1', 'Chromosome: 1; Location: 1p36.22', 'Chromosome: 21; Location: 21q22.3']\n",
      "['None', 'not_in_entrez', 'not_in_entrez', 'Annotation: Chromosome X NC_004354.4 (10233347..10263485, complement)', 'Annotation: Chromosome 3R NT_033777.3 (12061070..12062944, complement)', 'Annotation: Chromosome 2L NT_033779.5 (1724768..1729636, complement)', 'None', 'None', 'None', 'None', 'None', 'not_in_entrez', 'None', 'None', 'Annotation:  NC_000913.3 (2305108..2306754, complement)', 'Annotation:  NC_019063.1 (11..241)', 'not_in_entrez', 'Annotation:  NC_000913.3 (2906643..2907941, complement)', 'Annotation:  NC_000913.3 (3058666..3059325, complement)', 'None', 'Annotation:  NC_000913.3 (4159390..4160790, complement)', 'Annotation:  NC_000913.3 (2177512..2178564, complement)', 'None', 'Annotation:  NC_000913.3 (1806370..1807299)', 'None', 'Annotation: Chromosome X NC_000023.11 (154531390..154547586, complement)', 'None', 'Annotation: Chromosome 10 NC_000010.11 (69269991..69401882)', 'None', 'Annotation: Chromosome 1 NC_000001.11 (8861002..8878686, complement)', 'Annotation: Chromosome 2 NC_000002.12 (88691658..88750935)', 'None', 'Annotation: Chromosome 12 NC_000012.12 (124911645..124915041, complement)', 'Annotation: Chromosome 7 NC_000007.14 (6374495..6403967)', 'Annotation: Chromosome 1 NC_000001.11 (9234767..9271337)', 'Annotation: Chromosome 21 NC_000021.9 (44300034..44327381)']\n"
     ]
    }
   ],
   "source": [
    "# to get the chromosome, start, and end position and possibly other information from the gene entrez search\n",
    "# CAUTION: hasn't implemented through iteration yet\n",
    "from Bio import Entrez\n",
    "# function to separate lines of the gene search and fetch output txt file\n",
    "def parse_gene_entrez():\n",
    "    filename = \"temp_gene_entrez.txt\"\n",
    "    in_file = open(filename, \"rt\")\n",
    "    lines = []\n",
    "    # separate each line in the for for inspection and downstream analysis\n",
    "    for line in in_file:\n",
    "        lines.append(line.rstrip('\\n'))\n",
    "    return lines\n",
    "\n",
    "def find_loci(lines):\n",
    "    locus = \"not_in_entrez\"\n",
    "    for line in lines:\n",
    "        if \"Chromosome:\" in line:\n",
    "            locus = line\n",
    "            break\n",
    "    return locus\n",
    "\n",
    "def find_annotation(lines):\n",
    "    annotation = \"not_in_entrez\"\n",
    "    for line in lines:\n",
    "        if \"Annotation:\" in line:\n",
    "            annotation = line\n",
    "            break\n",
    "    return annotation\n",
    "\n",
    "def get_location(term_list):   \n",
    "    loci = []\n",
    "    annotations = []\n",
    "    Entrez.email = 'ych323@berkeley.edu'\n",
    "    for term in term_list:\n",
    "        data = search_fetch(\"gene\",term)\n",
    "        if data == \"None\":\n",
    "            locus = \"None\"\n",
    "            annotation = \"None\"\n",
    "        else:\n",
    "            f = open('temp_gene_entrez.txt','w')\n",
    "            f.write(data)\n",
    "            f.close()\n",
    "            lines = parse_gene_entrez()\n",
    "            # Location and annotation are in the 5 and 6 of the lines list \n",
    "            # if there's no desired field, then append \"insufficient info\"\n",
    "            locus = find_loci(lines)\n",
    "            annotation = find_annotation(lines)\n",
    "        loci.append(locus)\n",
    "        annotations.append(annotation)\n",
    "    return loci, annotations\n",
    "\n",
    "loci, annotations = get_location(term_list)\n",
    "\n",
    "print(loci)\n",
    "print(annotations)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome: X; Location: 1-31 cM\n",
      "Chromosome: 3R; Location: 3-51 cM\n",
      "Chromosome: 2L; Location: 2-3 cM\n",
      "Chromosome: X; Location: Xq28\n",
      "Chromosome: 10; Location: 10q22.1\n",
      "Chromosome: 1; Location: 1p36.23\n",
      "Chromosome: 2; Location: 2p11.2\n",
      "Chromosome: 12; Location: 12q24.31\n",
      "Chromosome: 7; Location: 7p22.1\n",
      "Chromosome: 1; Location: 1p36.22\n",
      "Chromosome: 21; Location: 21q22.3\n",
      "['None', 'unknown', 'unknown', ' X', ' 3R', ' 2L', 'None', 'None', 'None', 'None', 'None', 'unknown', 'None', 'None', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'None', ' X', 'None', ' 10', 'None', ' 1', ' 2', 'None', ' 12', ' 7', ' 1', ' 21']\n",
      "['None', 'unknown', 'unknown', ' 1-31 cM', ' 3-51 cM', ' 2-3 cM', 'None', 'None', 'None', 'None', 'None', 'unknown', 'None', 'None', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'not_in_entrez', 'None', 'not_in_entrez', 'None', ' Xq28', 'None', ' 10q22.1', 'None', ' 1p36.23', ' 2p11.2', 'None', ' 12q24.31', ' 7p22.1', ' 1p36.22', ' 21q22.3']\n"
     ]
    }
   ],
   "source": [
    "# parse the chromosome and location from loci list and the start and end position from the annotations site\n",
    "def parse_chromosome(loci):\n",
    "    chromosomes = []\n",
    "    locations = []\n",
    "    for temp_string in loci: \n",
    "        if \"Chromosome:\" not in temp_string:\n",
    "            location = temp_string\n",
    "            chromosome = temp_string\n",
    "        elif \"Un\" in temp_string:\n",
    "            location = \"unknown\"\n",
    "            chromosome = \"unknown\"\n",
    "        else:\n",
    "            print(temp_string)\n",
    "            temp_list = temp_string.split(\";\")\n",
    "            chromosome = temp_list[0].split(\":\")[1]\n",
    "            location = temp_list[1].split(\":\")[1]\n",
    "        locations.append(location)\n",
    "        chromosomes.append(chromosome)\n",
    "    return chromosomes, locations\n",
    "\n",
    "chromosomes, locations =  parse_chromosome(loci)\n",
    "\n",
    "print(chromosomes)\n",
    "print(locations)\n",
    "# from the list we can see that at lot of genes does not have info in all organisms\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotation: Chromosome X NC_004354.4 (10233347..10263485, complement)\n",
      "Annotation: Chromosome 3R NT_033777.3 (12061070..12062944, complement)\n",
      "Annotation: Chromosome 2L NT_033779.5 (1724768..1729636, complement)\n",
      "Annotation:  NC_000913.3 (2305108..2306754, complement)\n",
      "Annotation:  NC_019063.1 (11..241)\n",
      "Annotation:  NC_000913.3 (2906643..2907941, complement)\n",
      "Annotation:  NC_000913.3 (3058666..3059325, complement)\n",
      "Annotation:  NC_000913.3 (4159390..4160790, complement)\n",
      "Annotation:  NC_000913.3 (2177512..2178564, complement)\n",
      "Annotation:  NC_000913.3 (1806370..1807299)\n",
      "Annotation: Chromosome X NC_000023.11 (154531390..154547586, complement)\n",
      "Annotation: Chromosome 10 NC_000010.11 (69269991..69401882)\n",
      "Annotation: Chromosome 1 NC_000001.11 (8861002..8878686, complement)\n",
      "Annotation: Chromosome 2 NC_000002.12 (88691658..88750935)\n",
      "Annotation: Chromosome 12 NC_000012.12 (124911645..124915041, complement)\n",
      "Annotation: Chromosome 7 NC_000007.14 (6374495..6403967)\n",
      "Annotation: Chromosome 1 NC_000001.11 (9234767..9271337)\n",
      "Annotation: Chromosome 21 NC_000021.9 (44300034..44327381)\n",
      "['None', 'not_in_entrez', 'not_in_entrez', '10233347', '12061070', '1724768', 'None', 'None', 'None', 'None', 'None', 'not_in_entrez', 'None', 'None', '2305108', '11', 'not_in_entrez', '2906643', '3058666', 'None', '4159390', '2177512', 'None', '1806370', 'None', '154531390', 'None', '69269991', 'None', '8861002', '88691658', 'None', '124911645', '6374495', '9234767', '44300034']\n",
      "['None', 'not_in_entrez', 'not_in_entrez', '10263485', '12062944', '1729636', 'None', 'None', 'None', 'None', 'None', 'not_in_entrez', 'None', 'None', '2306754', '241', 'not_in_entrez', '2907941', '3059325', 'None', '4160790', '2178564', 'None', '1807299', 'None', '154547586', 'None', '69401882', 'None', '8878686', '88750935', 'None', '124915041', '6403967', '9271337', '44327381']\n"
     ]
    }
   ],
   "source": [
    "def parse_annotation(annotations):\n",
    "    starts = []\n",
    "    ends = []\n",
    "    for temp_string in annotations: \n",
    "        if \"Annotation:\" not in temp_string:\n",
    "            start = temp_string\n",
    "            end = temp_string\n",
    "        elif \"Un\" in temp_string:\n",
    "            start = temp_string\n",
    "            end = temp_string\n",
    "        else:\n",
    "            print(temp_string)\n",
    "            a = temp_string\n",
    "            b = a.split(\"(\")\n",
    "            c = b[-1].split(\")\")\n",
    "            d = c[0].split(\" \")\n",
    "            start_end = d[0].split(\"..\")\n",
    "            start = start_end[0]\n",
    "            end = start_end[1].split(\",\")[0]\n",
    "        starts.append(start)\n",
    "        ends.append(end)\n",
    "    return starts, ends\n",
    "\n",
    "starts, ends = parse_annotation(annotations)\n",
    "print(starts)\n",
    "print(ends)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sanity check on the entrez gene output\n",
    "> as we have all the lists for the gene table insertion, check that the length of the lists are consistent\n",
    "\n",
    "- id INTEGER PRIMARY KEY  \n",
    "- name TEXT  \n",
    "- description TEXT  \n",
    "- organism TEXT  \n",
    "- nucleotide_sequence TEXT  \n",
    "- chromosome TEXT  \n",
    "- location TEXT  \n",
    "- start INT  \n",
    "- end INT  \n",
    "- strand VARCHAR(1)  \n",
    "- translated_sequence TEXT  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n",
      "36\n"
     ]
    }
   ],
   "source": [
    "print(len(name_list))\n",
    "print(len(organism_list))\n",
    "print(len(descriptions_list))\n",
    "print(len(strand_list))\n",
    "print(len(nucleotide_list))\n",
    "print(len(protein_list))\n",
    "print(len(chromosomes))\n",
    "print(len(locations))\n",
    "print(len(starts))\n",
    "print(len(ends))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO genes (name,description,organism, nucleotide_sequence,chromosome,location,start,end,strand,translated_sequence) VALUES(?,?,?,?,?,?,?,?,?,?)'''\n",
    "\n",
    "# insertion\n",
    "for i in range(len(name_list)):\n",
    "    name = name_list[i]\n",
    "    description = descriptions_list[i]\n",
    "    organism = organism_list[i]\n",
    "    nucleotide_sequence = nucleotide_list[i]\n",
    "    chromosome = chromosomes[i]\n",
    "    location = locations[i]\n",
    "    start = starts[i]\n",
    "    end = ends[i]\n",
    "    strand = strand_list[i]\n",
    "    translated_sequence = protein_list[i]\n",
    "    temp_value = (name,description,organism, nucleotide_sequence,chromosome,location,start,end,strand,translated_sequence)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Demo of the functionality of three individual tables\n",
    "Special Place Holder Explanation:  \n",
    "- None: The entrez search has an empty search result  \n",
    "- not in entrez: even if the search leads to output, the gene annotation information does not contain the corresponding info   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
