{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Procedural Notes  \n",
    "\n",
    "### Enzyme/Pathway Selection  \n",
    "Kegg glycolysis/Glucogenesis-Reference Pathway:  \n",
    "\n",
    ">Enzymes selected for each pathway are:  \n",
    ">- Glycolysis: K00844, K11645, K01689, K16370  \n",
    ">- Citric Acid Cycle: K01647, K00030, K01900, K00116  \n",
    ">- Pentose Phosphate Pathway: K13937, K00036, K01807, K06859  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating the genes, pathways and enzymes table with respective assigned columns\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "# to avoid conflict with previously built table\n",
    "c.execute(\"\"\"DROP TABLE genes\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE pathways\"\"\")\n",
    "c.execute(\"\"\"DROP TABLE enzymes\"\"\")\n",
    "\n",
    "# Caution: use INTEGER instead of INT for primary key generation\n",
    "c.execute(\"\"\"CREATE TABLE genes (id INTEGER PRIMARY KEY, name TEXT, description TEXT, organism TEXT, nucleotide_sequence TEXT\n",
    "chromosome TEXT, start INT, end INT, strand VARCHAR(1), translated_sequence TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathways (id INTEGER PRIMARY KEY, name TEXT, description TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzymes (id INTEGER PRIMARY KEY, kegg_id TEXT, name TEXT, function TEXT, ec_number TEXT);\"\"\")\n",
    "\n",
    "conn.commit()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inserting values into the pathway table\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO pathways (name, description) VALUES(?,?)'''\n",
    "\n",
    "names = [\"Glycolysis\",\"Citric Acid Cycle\",\"Pentose Phosphate\"]\n",
    "descriptions = [\"the breakdown of glucose by enzymes, releasing energy and pyruvic acid.\" , \"the sequence of reactions by which most living cells generate energy during the process of aerobic respiration. It takes place in the mitochondria, consuming oxygen, producing carbon dioxide and water as waste products, and converting ADP to energy-rich ATP\",\"a metabolic pathway parallel to glycolysis, It generates NADPH and pentoses (5-carbon sugars) as well as ribose 5-phosphate, the last one a precursor for the synthesis of nucleotides\"]\n",
    "\n",
    "\n",
    "# Take in two list with equal length and insert the into the corresponding field\n",
    "for i in range(len(names)):\n",
    "    name = names[i]\n",
    "    description = descriptions[i]\n",
    "    temp_value = (name,description)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query the name and ec_number of the selected Enzymes from KEGG\n",
    "1. Make a sample.list file in the current directory with the following lines  \n",
    "K00844   \n",
    "K11645   \n",
    "K01689  \n",
    "K16370  \n",
    "K01647  \n",
    "K00030  \n",
    "K01900  \n",
    "K00116  \n",
    "K13937  \n",
    "K00036  \n",
    "K01807  \n",
    "K06859  \n",
    "\n",
    "2. In the terminal execute: `curl -g -s -S http://rest.kegg.jp/list/ko | grep -f sample.list | sed \"s/cpd\\://\" > sample_extracted.table.txt`\n",
    "\n",
    "3. Inspect the output file sample_extracted.table.txt  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['ko:K00030',\n",
       "  'ko:K00036',\n",
       "  'ko:K00116',\n",
       "  'ko:K00844',\n",
       "  'ko:K01647',\n",
       "  'ko:K01689',\n",
       "  'ko:K01807',\n",
       "  'ko:K01900',\n",
       "  'ko:K06859',\n",
       "  'ko:K11645',\n",
       "  'ko:K13937',\n",
       "  'ko:K16370'],\n",
       " ['IDH3',\n",
       "  'G6PD,zwf',\n",
       "  'mqo',\n",
       "  'HK',\n",
       "  'CS,gltA',\n",
       "  'ENO,eno',\n",
       "  'rpiA',\n",
       "  'LSC2',\n",
       "  'pgi1',\n",
       "  'fbaB',\n",
       "  'H6PD',\n",
       "  'pfkB'],\n",
       " ['isocitrate dehydrogenase (NAD+)',\n",
       "  'glucose-6-phosphate 1-dehydrogenase',\n",
       "  'malate dehydrogenase (quinone)',\n",
       "  'hexokinase',\n",
       "  'citrate synthase',\n",
       "  'enolase',\n",
       "  'ribose 5-phosphate isomerase A',\n",
       "  'succinyl-CoA synthetase beta subunit',\n",
       "  'glucose-6-phosphate isomerase, archaeal',\n",
       "  'fructose-bisphosphate aldolase, class I',\n",
       "  'hexose-6-phosphate dehydrogenase',\n",
       "  '6-phosphofructokinase 2'],\n",
       " ['EC:1.1.1.41]',\n",
       "  'EC:1.1.1.49 1.1.1.363]',\n",
       "  'EC:1.1.5.4]',\n",
       "  'EC:2.7.1.1]',\n",
       "  'EC:2.3.3.1]',\n",
       "  'EC:4.2.1.11]',\n",
       "  'EC:5.3.1.6]',\n",
       "  'EC:6.2.1.4 6.2.1.5]',\n",
       "  'EC:5.3.1.9]',\n",
       "  'EC:4.1.2.13]',\n",
       "  'EC:1.1.1.47 3.1.1.31]',\n",
       "  'EC:2.7.1.11]'])"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parse in the sample_extracted.table.txt for enzymes table insertion\n",
    "def parse_extracted_table():\n",
    "    filename = \"sample_extracted.table.txt\"\n",
    "    in_file = open(filename, \"rt\")\n",
    "    lines = []\n",
    "    for line in in_file:\n",
    "        lines.append(line.rstrip('\\n'))\n",
    "        print(line)\n",
    "    return lines\n",
    "\n",
    "lines = parse_extracted_table()\n",
    "\n",
    "\n",
    "\n",
    "def get_values(lines):\n",
    "    kegg_ids = []\n",
    "    names = []\n",
    "    functions = []\n",
    "    ec_numbers = []\n",
    "    for line in lines:\n",
    "        line_ec = line.split(\"[\")\n",
    "        ec_number = line_ec[-1]\n",
    "        line_2_part = line_ec[0].split(\";\")\n",
    "        line_part1 = line_2_part[0].split()\n",
    "        line_part2 = line_2_part[1].split()\n",
    "#         print(line_part1)\n",
    "#         print(line_part2)\n",
    "        kegg_id = line_part1[0]\n",
    "        temp_name = line_part1[1:]\n",
    "        temp_name[:] = [''.join(temp_name[:])]\n",
    "        name = temp_name[0]\n",
    "        temp_function = line_part2[:]\n",
    "        temp_function[:] = [\" \".join(temp_function[:])]\n",
    "        function = temp_function[0]     \n",
    "        kegg_ids.append(kegg_id)\n",
    "        names.append(name)\n",
    "        functions.append(function)\n",
    "        ec_numbers.append(ec_number)\n",
    "    return kegg_ids,names,functions,ec_numbers\n",
    "\n",
    "get_values(lines)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert values into the enzymes table\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "cmd = ''' INSERT INTO enzymes (kegg_id, name, function, ec_number) VALUES(?,?,?,?)'''\n",
    "\n",
    "\n",
    "kegg_ids,names,functions,ec_numbers = get_values(lines)\n",
    "for i in range(len(names)):\n",
    "    name = names[i]\n",
    "    function = functions[i]\n",
    "    kegg_id = kegg_ids[i]\n",
    "    ec_number = ec_numbers[i]\n",
    "    temp_value = (kegg_id, name, function, ec_number)\n",
    "    c.execute(cmd,temp_value)\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DictElement({'Count': '303', 'RetMax': '20', 'RetStart': '0', 'IdList': ['L44140.1', 'KJ896841.1', 'S64462.1', 'S58359.1', 'NM_000402.4', 'NM_001360016.1', 'NM_001042351.2', 'AB376963.1', 'X55448.1', 'DQ173568.1', 'DQ839546.1', 'DQ832766.1', 'M12996.1', 'MG772799.1', 'DQ173642.1', 'DQ173641.1', 'DQ173640.1', 'DQ173639.1', 'DQ173638.1', 'DQ173637.1'], 'TranslationSet': [DictElement({'From': 'homo sapiens[ORGN]', 'To': '\"Homo sapiens\"[Organism]'}, attributes={})], 'TranslationStack': [DictElement({'Term': '\"Homo sapiens\"[Organism]', 'Field': 'Organism', 'Count': '16629335', 'Explode': 'Y'}, attributes={}), DictElement({'Term': 'G6PD[All Fields]', 'Field': 'All Fields', 'Count': '41938', 'Explode': 'N'}, attributes={}), 'AND'], 'QueryTranslation': '\"Homo sapiens\"[Organism] AND G6PD[All Fields]'}, attributes={})\n",
      ">KJ896841.1 Synthetic construct Homo sapiens clone ccsbBroadEn_06235 G6PD gene, encodes complete protein\n",
      "GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAACTTTGTACAAAAAAGTTGGCATGGC\n",
      "AGAGCAGGTGGCCCTGAGCCGGACCCAGGTGTGCGGGATCCTGCGGGAAGAGCTTTTCCAGGGCGATGCC\n",
      "TTCCATCAGTCGGATACACACATATTCATCATCATGGGTGCATCGGGTGACCTGGCCAAGAAGAAGATCT\n",
      "ACCCCACCATCTGGTGGCTGTTCCGGGATGGCCTTCTGCCCGAAAACACCTTCATCGTGGGCTATGCCCG\n",
      "TTCCCGCCTCACAGTGGCTGACATCCGCAAACAGAGTGAGCCCTTCTTCAAGGCCACCCCAGAGGAGAAG\n",
      "CTCAAGCTGGAGGACTTCTTTGCCCGCAACTCCTATGTGGCTGGCCAGTACGATGATGCAGCCTCCTACC\n",
      "AGCGCCTCAACAGCCACATGAATGCCCTCCACCTGGGGTCACAGGCCAACCGCCTCTTCTACCTGGCCTT\n",
      "GCCCCCGACCGTCTACGAGGCCGTCACCAAGAACATTCACGAGTCCTGCATGAGCCAGATAGGCTGGAAC\n",
      "CGCATCATCGTGGAGAAGCCCTTCGGGAGGGACCTGCAGAGCTCTGACCGGCTGTCCAACCACATCTCCT\n",
      "CCCTGTTCCGTGAGGACCAGATCTACCGCATCGACCACTACCTGGGCAAGGAGATGGTGCAGAACCTCAT\n",
      "GGTGCTGAGATTTGCCAACAGGATCTTCGGCCCCATCTGGAACCGGGACAACATCGCCTGCGTTATCCTC\n",
      "ACCTTCAAGGAGCCCTTTGGCACTGAGGGTCGCGGGGGCTATTTCGATGAATTTGGGATCATCCGGGACG\n",
      "TGATGCAGAACCACCTACTGCAGATGCTGTGTCTGGTGGCCATGGAGAAGCCCGCCTCCACCAACTCAGA\n",
      "TGACGTCCGTGATGAGAAGGTCAAGGTGTTGAAATGCATCTCAGAGGTGCAGGCCAACAATGTGGTCCTG\n",
      "GGCCAGTACGTGGGGAACCCCGATGGAGAGGGCGAGGCCACCAAAGGGTACCTGGACGACCCCACGGTGC\n",
      "CCCGCGGGTCCACCACCGCCACTTTTGCAGCCGTCGTCCTCTATGTGGAGAATGAGAGGTGGGATGGGGT\n",
      "GCCCTTCATCCTGCGCTGCGGCAAGGCCCTGAACGAGCGCAAGGCCGAGGTGAGGCTGCAGTTCCATGAT\n",
      "GTGGCCGGCGACATCTTCCACCAGCAGTGCAAGCGCAACGAGCTGGTGATCCGCGTGCAGCCCAACGAGG\n",
      "CCGTGTACACCAAGATGATGACCAAGAAGCCGGGCATGTTCTTCAACCCCGAGGAGTCGGAGCTGGACCT\n",
      "GACCTACGGCAACAGATACAAGAACGTGAAGCTCCCTGACGCCTATGAGCGCCTCATCCTGGACGTCTTC\n",
      "TGCGGGAGCCAGATGCACTTCGTGCGCAGCGACGAGCTCCGTGAGGCCTGGCGTATTTTCACCCCACTGC\n",
      "TGCACCAGATTGAGCTGGAGAAGCCCAAGCCCATCCCCTATATTTATGGCAGCCGAGGCCCCACGGAGGC\n",
      "AGACGAGCTGATGAAGAGAGTGGGTTTCCAGTATGAGGGCACCTACAAGTGGGTGAACCCCCACAAGCTC\n",
      "TGCCCAACTTTCTTGTACAAAGTTGGCATTATAAGAAAGCATTGCTTATCAATTTGTTGCAACGAAC\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "#Entrez.esearch(db,term, sort)\n",
    "#Entrez.efetch(db, id, rettype, retmode)\n",
    "#for enzymes and genes only; Entrez doesn't have pathway info\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Entrez.email = 'ych323@berkeley.edu'\n",
    "handle = Entrez.esearch(db = 'nucleotide',\n",
    "                       term = 'homo sapiens[ORGN] G6PD',\n",
    "                       sort= 'relevance',\n",
    "                       idtype= 'acc')\n",
    "\n",
    "fetched_dict = Entrez.read(handle)\n",
    "print(fetched_dict)\n",
    "\n",
    "handle = Entrez.efetch(db = 'nucleotide', id = fetched_dict[\"IdList\"][1], rettype = 'fasta', retmode = 'text')\n",
    "print(handle.read())\n",
    "# TODO: Make a function to iterate through the query\n",
    "# for i in Entrez.read(handle)[\"IdList\"]:\n",
    "#     handle = Entrez.efetch(db = 'nucleotide', id = i, rettype = 'fasta', retmode = 'text')\n",
    "#     print(handle.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
