{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary of the Terminal commands used ###\n",
    "\n",
    "__mkdir lab02__: make directory named lab02  \n",
    "__cd lab02__: change directory to new lab02 directory  \n",
    "__mkdir src__: make directory src (source code)  \n",
    "__mkdir data__: make directory data (to store data)  \n",
    "__cd data__: enter data directory  \n",
    "#### used jupyter notebook to upload seqs.fasta####\n",
    "__muscle -in seqs.fa -out seqs.aligned.fa__ : call muscle command given seqs.fa as input; output the aligned version of the sequences as seqs.aligned.fa  \n",
    "__more seqs.fa__ & __more seqs.aligned.fa__ : 2 separate commands to view differences between input and output files  \n",
    "__fasttree â€“nt seqs.aligned.fa > tree.nwk__ : generate a newick-formatted tree\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Phylo\n",
    "from io import StringIO\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "\n",
    "# read in the tree object\n",
    "tree = Phylo.read(\"tree.nwk\", \"newick\")\n",
    "# set the font size and figure size\n",
    "mpl.rcParams.update({'font.size': 15})\n",
    "plt.rcParams['figure.figsize'] = [60, 40]\n",
    "#plot the graph\n",
    "Phylo.draw(tree)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "my_seqs = [];\n",
    "obvious_clusters = [\"pi.3\", \"bb.2\", \"hu.40\", \"cy.6\", \"rh.36\", \"rh.51\", \"rh.48\", \"hu.48\", \"hu.31\"];\n",
    "for seq in SeqIO.parse(\"seqs.fa\", \"fasta\"):\n",
    "    if seq.id in obvious_clusters:\n",
    "        my_seqs.append(seq)\n",
    "for seq in my_seqs:\n",
    "    print(seq.id + \":\")\n",
    "    print(seq.seq)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__BLAST__  \n",
    "\n",
    "bb.2 : Non-human primate Adeno-associated virus isolate AAVbb.2 capsid protein (VP1) gene, complete cds   \n",
    "\n",
    "cy.6 : Non-human primate Adeno-associated virus isolate AAVcy.6 capsid protein (VP1) gene, complete cds  \n",
    "\n",
    "hu.31:Adeno-associated virus isolate hu.31 capsid protein VP1 (cap) gene, complete cds  \n",
    "\n",
    "hu.40: Adeno-associated virus isolate hu.40 capsid protein VP1 (cap) gene, complete cds  \n",
    "\n",
    "hu.48: Adeno-associated virus isolate hu.48 capsid protein VP1 (cap) gene, complete cds  \n",
    "\n",
    "pi.3:  Adeno-associated virus isolate pi.3 capsid protein VP1 (cap) gene, complete cds  \n",
    "\n",
    "rh.36: Non-human primate Adeno-associated virus isolate AAVrh.36 capsid protein (VP1) gene, complete cds  \n",
    "\n",
    "rh.48: Adeno-associated virus isolate rh.48 capsid protein VP1 (cap) gene, complete cds  \n",
    "\n",
    "rh.51: Adeno-associated virus isolate rh.51 capsid protein VP1 (cap) gene, complete cds  \n",
    "\n",
    "__What are some of the most common matches for each cluster? Write them down in your lab notebook. Do they all match the same gene? Are there any that definitely do not match the others? Why might we not trust the annotations for sequences that come up in our BLAST?__  \n",
    "\n",
    "The most common match for each cluster follows the same format as those listed above, with some variation in the particular capsid proteins isolated. They all seem to match the VP1 gene. As mentioned in lecture, it's important to note that a lot of databases are inherently biased in the information they store, based on the historical trends in the gathering and publishing of data.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "cluster1 = [\"rh.40\", \"hu.67\", \"hu.40\", \"hu.37\", \"hu.66\", \"hu.38\", \"hu.41\", \"hu.42\"]\n",
    "cluster2 = [\"rh.10\", \"bb.1\", \"hu.6\", \"bb.2\", \"hu.17\", \"rh.2\"]\n",
    "cluster3 = [\"pi.2\", \"pi.3\", \"pi.1\"]\n",
    "cluster4 = [\"hu.39\", \"rh.50\", \"rh.51\", \"rh.49\", \"rh.57\", \"rh.61\", \"rh.64\", \"rh.58\", \"rh.53\"]\n",
    "cluster5 = [\"cy.6\", \"cy.3\", \"cy.4\", \"cy.5\", \"rh.13\"]\n",
    "cluster6 = [\"rh.35\", \"rh.36\", \"rh.37\"]\n",
    "cluster7 = [\"rh.48\", \"rh.55\", \"rh.62\"]\n",
    "cluster8 = [\"hu.14\", \"hu.31\", \"hu.38\"]\n",
    "def clusterseq_maker(cluster):\n",
    "    my_seqs =[];\n",
    "    for seq in SeqIO.parse(\"seqs.fa\", \"fasta\"):\n",
    "        if seq.id in cluster:\n",
    "            my_seqs.append(SeqRecord(seq.seq))\n",
    "    return my_seqs\n",
    "\n",
    "cluster1seqs =clusterseq_maker(cluster1)\n",
    "cluster2seqs=clusterseq_maker(cluster2)\n",
    "cluster3seqs=clusterseq_maker(cluster3)\n",
    "cluster4seqs= clusterseq_maker(cluster4)\n",
    "cluster5seqs =clusterseq_maker(cluster5)\n",
    "cluster6seqs =clusterseq_maker(cluster6)\n",
    "cluster7seqs =clusterseq_maker(cluster7)\n",
    "cluster8seqs =clusterseq_maker(cluster8)\n",
    "\n",
    "def msa_maker(clusterseqs):\n",
    "    return MultipleSeqAlignment(clusterseqs)\n",
    "\n",
    "msa1 = msa_maker(cluster1seqs)\n",
    "msa2 = msa_maker(cluster2seqs)\n",
    "msa3 = msa_maker(cluster3seqs)\n",
    "msa4= msa_maker(cluster4seqs)\n",
    "msa5 = msa_maker(cluster5seqs)\n",
    "msa6 = msa_maker(cluster6seqs)\n",
    "msa7 = msa_maker(cluster7seqs)\n",
    "msa8 = msa_maker(cluster8seqs)\n",
    "\n",
    "print(cluster1seqs)\n",
    "print(msa1)\n",
    "\n",
    "def atgc_counter(msa):\n",
    "    atcount= []\n",
    "    gccount=[]\n",
    "    i = 0\n",
    "    while i< len(msa[0]):\n",
    "        atcount.append(0)\n",
    "        gccount.append(0)\n",
    "    \n",
    "    i = 0;\n",
    "    while i < len(msa[0]):\n",
    "        curr = msa[:,i]\n",
    "        print(curr)\n",
    "        for letter in curr:\n",
    "            if letter == \"a\" or letter == \"t\":\n",
    "                atcount[i] +=1\n",
    "            else:\n",
    "                gccount[i] +=1\n",
    "        i +=1\n",
    "    return atcount, gccount;\n",
    "\n",
    "at1,gc1 = atgc_counter(msa1)\n",
    "print(at1)\n",
    "print(gc1)           \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
